KEP-4212: Improve Node Maintenance (#4213)
kubernetes/enhancements: keps/sig-apps/4212-declarative-node-maintenance/README.md

and integrated with multiple components to observe the benefits. And those components are both
admin and application developer facing.

There is a [Node Maintenance Operator](https://github.com/medik8s/node-maintenance-operator) project
that provides a similar API and has some adoption. But it is not at a level where applications could
depend on this API to be present in the cluster. So it doesn't make that much sense to implement
the app migration logic as it cannot be applied everywhere. As shown in the motivation part
of this KEP, there is a big appetite for having a unified and stable API that could be used by
everyone to implement the new capabilities.

### Use a Node Object Instead of Introducing a New NodeMaintenance API
As an alternative, it would be possible to signal the node maintenance by marking the node object
instead of introducing a new API. But it is probably better to decouple this from the node for
extensibility reasons. As we can see, the `kubectl drain` logic is a bit complex, and it may be
possible to move this logic to a controller in the future and make the node maintenance purely
declarative. It also helps to decouple RBAC permissions from the node object.

> 

Reviewer: Another benefit IMO: you could have two or more different grounds for node maintenance, for a given node, that overlap in time.

Author: +1, I have added this point there

   `.spec.terminationGracePeriodSeconds` of each pod and may cause premature termination of
   the application. This has been discussed in issue [kubernetes/kubernetes#75482](https://github.com/kubernetes/kubernetes/issues/75482)
   and in issue [kubernetes-sigs/cluster-api#6158](https://github.com/kubernetes-sigs/cluster-api/issues/6158).
4. There are cases during a node shutdown, when data corruption can occur due to premature node
   shutdown. It would be great if applications could perform data migration and synchronization of 
   cached writes to the underlying storage before the pod deletion occurs. This is not easy to
   quantify even with pod's `.spec.shutdownGracePeriod`, as the time depends on the size of the data
   and the speed of the storage. This has been discussed in issue [kubernetes/kubernetes#116618](https://github.com/kubernetes/kubernetes/issues/116618)
   and in issue [kubernetes/kubernetes#115148](https://github.com/kubernetes/kubernetes/issues/115148).
5. There is not enough metadata about why the node drain was requested. This has been discussed in
   issue [kubernetes/kubernetes#30586](https://github.com/kubernetes/kubernetes/issues/30586).

> Approaches and workarounds used by other projects to deal with these shortcomings:

Reviewer 2: Is https://github.com/kubereboot/kured  relevant to mention here?

Author: Thanks for highlighting another example. It seems kured is using the kubectl drain helper to achieve the drain.

Author: I have mentioned the kured project in the examples.

- Introduce a field that could include non-critical daemon set pods
  (priority `system-cluster-critical` or `system-node-critical`) for node maintenance/drain request.
  The daemon set controller would then gracefully shut down these pods. Critical pods could be
  overridden by the priority list mentioned below.
- NodeMaintenance could include a plan of which pods to target first. Similar to graceful node
  shutdown, we could include a list of priorities to decide which pods should be terminated first.
  This list could optionally include pod timeouts, but could also wait for all the pods of a given
  priority class to finish first without a timeout. This could also be used to target daemon set
  pods of certain priorities (see point above). We could also introduce drain profiles based on
  these lists. The cluster admin could then choose or create such a profile based on his/her needs.
  The logic for processing the decision list would be contained in the node maintenance controller,
  which would set an intent to selected pods to shut down via the TargetedByNodeMaintenance
  condition.

> - Introduce a node maintenance period or a TTL optional field as an upper bound on the duration of

Author: This might not be  a reliable way of delivering that information and there could be other better mechanism. But, if there is a sufficient interest, I think it can be added in later (for example suspended field in the spec).

> - Introduce a node maintenance period or a TTL optional field as an upper bound on the duration of

Author: I have changed the NodeMaintenance API: now it should be possible to create the object without starting the actual maintenance. This can be used to announce the upcoming maintenance via `.spec.reason`.

  schedulable again.

## Proposal
Most of these issues stem from missing a standardized way of detecting a start of the node drain.
This KEP proposes the introduction of a NodeMaintenance object that would signal an intent to
gracefully remove pods from a given node. This should complement the node's `.spec.unschedulable`,
which prevents new pods from being scheduled on such a node. We will focus primarily on
`kubectl drain`, but the API can also be used by other drain implementations as well (e.g. node
autoscalers).

We will first introduce the NodeMaintenance API and then later modify the behavior of the
Kubernetes system to fix all the node drain issues we mentioned earlier. The name, NodeMaintenance
is subject to change; please see the alternatives section for other suggestions.

To support workload migration, a new controller should be introduced to observe the NodeMaintenance
objects and then mark pods for migration or removal with a condition. The pods would be selected
according to the node at first, but the selection mechanism can be extended later. Controllers can
then implement the migration. The advantage of this approach is that controllers do not have to be

> aware of the NodeMaintenance object (no RBAC changes required). They only have to observe pods they

Reviewer 3: Presumably the node would be marked in some fashion as well?

Author: I am not sure yet if and how we should do this (condition, annotation). Are you aware of any potential consumers for this?

Reviewer 3: Yes. I had a conversation with a customer right before reading this where the centralized dev ops team did exactly that as a mechanism to notify the application teams, acking that it's not a great mechanism, but it's something that can be automated around. From a cluster admin perspective, it allows you to provide that automation to notify and whatnot.

Author: I have added a mention under the node maintenance controller about marking the node.

drain a node using the same API and provide a similar experience to the CLI counterpart.

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

### Kubectl

`kubectl drain` command will be changed to create a NodeMaintenance object after marking node
unschedulable. We will also change the implementation to skip applications that support
workload migration. At first, we will just check deployments with a positive
`.spec.strategy.rollingUpdate.maxSurge` value. To make this extensible, we propose a new Surge API
that could be checked before disrupting any workload (see below). If the cluster,

> doesn't support the NodeMaintenance API, kubectl will perform the node drain in a backwards

Reviewer 3: Will `kubectl` fail if you set `--node-maintenance` and it doesn't actually support that? Seems like it should in that case so the operator knows.

Author: We need to make `kubectl drain` compatible with the old behaviour, so we cannot fail there in this scenario. I think I will remove the `--node-maintenance` option from `kubectl uncordon` for now, as it doesn't make much sense in case the selected nodes do not overlap with the `NodeMaintenance` object.

Author: In the current version, NodeMaintenance object and controller is taking a full responsibility for cordoning and uncordoning the nodes. The kubectl commands should not be necessary anymore. 

I like the idea of showing warnings when users use `kubectl cordon` and `kubectl uncordon`. These commands will become noop as the node maintenance controller will stomp over their changes.

compatible way.

`kubectl uncordon` will be amended with `--node-maintenance` option. This can be used to delete the
NodeMaintenance object together with marking the node as schedulable again. The default value of
this flag will be `false`.

> ### NodeMaintenance API

Author: added a `.spec.nodeSelector` to the API

NodeMaintenance objects serve as an intent to remove or migrate pods from a particular node. We will
also support setting a reason for the maintenance.

```golang

type NodeMaintenance struct {
    ...
    Spec NodeMaintenanceSpec
    Status NodeMaintenanceStatus
}

type NodeMaintenanceSpec struct {
    NodeName string
    Reason string
}

type NodeMaintenanceStatus struct {
    // number of pods this node maintenance is requesting to terminate

>     TargetedPods int32

Author: Since we do not target all pods, `RemainingPods` doesn't feel right. But `PodsPendingTermination` could be workable.

>     TargetedPods int32

Reviewer 3: Ya. I think just keeping it as explicit as possible so people don't need to read the documentation to know what it means.

>     TargetedPods int32

Author: I have started using a new term, Evacuation, which I think describes the desired meaning pretty well. Now the status is per node and has two fields: `PodsPendingEvacuation` and `PodsEvacuating`.

}
```

The NodeMaintenance API is being introduced with a minimal set of features, but it is expected that

> additional features will be added to the API in the future.

Reviewer: Not needed for this PR; I'll mention it anyway in case it's useful.

Sometimes people want node maintenance to trigger a drain; other times, just a cordon.

Author: You mean cordon in a sense that makes the node unschedulable? I am not sure if this should be part of the API if users can just use `kubectl cordon` or edit nodes manually.

Reviewer: **Future use**

One NodeMaintenance with a selector could cordon all nodes with some property (eg: all the AArch64 nodes that also have `feature.node.kubernetes.io/example-thing` labels).

Author: I have updated the KEP to include the phases of node maintenance. One of them is cordon. These phases can now be toggled with the NodeMaintenance API.

The design of the API should not assume that only one node is selected for the node maintenance.
And a Node to NodeMaintenance mapping should not be used or encouraged. `.spec.nodeName` field
could be made optional if the requirements change in the future, in favor of `.spec.nodeSelector`.
This could cause pods on multiple nodes to be marked for the maintenance at the same time, so even
the NodeMaintenance `.status` should be made general enough.

<<[UNRESOLVED NodeMaintenance ]>>

- It is unclear whether we should allow marking the NodeMaintenances as finished and keeping the
  old maintenances in the system for bookkeeping purposes.

<<[/UNRESOLVED]>>

### NodeMaintenance Controller

Node maintenance controller will be introduced and added to `kube-controller-manager`. It will
observe NodeMaintenance objects and set `TargetedByNodeMaintenance` condition on selected pods.
The pods would be selected according to the node. Controllers that own these pods, would observe

> them and start a removal or migration from these nodes upon detecting this condition.

Reviewer 4: The ordering of pod conditions might matter to some operators. Take an operator that updates on a per failure domain for example (eg: https://github.com/kubernetes/kubernetes/issues/114010). Ideally the operator would get a signal to migrate a subset of pods at the same time, so a migration away from affected nodes could be done strategically.

I can't think of a better messaging mechanism than conditions on pods however, so this may just need to be a design consideration of any controller/operator that is using this new condition.

Author: Thanks, we can take that into consideration, but should depend on it. API conventions specify that [conditions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties)

> This collection should be treated as a map with a key of type

The operator can also observe and parse the NodeMaintenance object as well and/or wait for the pods to be marked. 

We can keep the the discussion open, how to achieve the best messaging mechanism.

Author: We have moved from the pod conditions to the dedicated API: https://github.com/kubernetes/enhancements/pull/4565

The selection process can be later enhanced to include priority or pod type (to target daemon sets
which we do not currently support).

The controller would also remove the `TargetedByNodeMaintenance` condition from the targeted pods
upon premature `NodeMaintenance` object removal.

The controller can then show progress by reconciling `.status.targetedPods`, to show how many pods
remain to be removed from the node.

The controller would also emit a warning event if the targeted node was not marked as unschedulable.
We could prevent this, by stomping on any changes to the node's `.spec.unschedulable` field and set
it to true for the whole lifetime of `NodeMaintenance` object.

<<[UNRESOLVED NodeMaintenance .status ]>>

> - Should we target all pods or just the ones we know can be deleted (drain filters)? The later will

Author: > Ideally, we'd target all pods that we'd like to be removed.

knowing which pods we (i.e. the admin) would like to be removed is the difficult part. Since we would like to automate the more difficult cases (drain filters) it would be hard to do drastic actions in the future if you cannot opt out from them. That is, if we mark them now, and removal feature is implemented in the controllers later, we could break users.

To support the observability (manual actions) we could just count them in the status instead for now (e.g. `NonTargetedPods`).

> - Should we target all pods or just the ones we know can be deleted (drain filters)? The later will

Author: I have tried to include most of the drain filters and add a reasoning why we can include them. Nevertheless I think we still have to skip DaemonSet pods as they can have critical services and should be opted in.

Would be great to get feedback about the pods with local storage though.

are blocking cluster upgrade. If the `.status.disruptionsAllowed` equals to 0, because for example
`.spec.maxUnavailable` is equal to 0, the cluster admin can assess that the PDB will not block
cluster upgrade if he or she observes `.status.expectedSurgePods` greater than 0.

### Surge API (Optional)

Any CustomResourceDefinition or built-in type could implement a surge API. This could be a
subresource similar to a scale API, that would serve for advertising the surge support. If a
certain workload supports a pod surge, then one of `.status.availableSurgeReplicas` or
`.status.currentSurgeReplicas` values will be positive.

```golang
type Surge struct {
    ...
    Status SurgeStatus
}

type SurgeStatus struct {
    // number of replicas that can surge during an application rollout or node maintenance
    SurgeReplicas int32
    // current number of surge replicas that are running above the normal amount of replicas
    CurrentSurgeReplicas int32
}

```

This API would be then used by kubectl and disruption controller to reason about the capabilities
of an application. It would be similar to how we use the scale API in podControllerFinders in

> disruption controller.

Reviewer 4: Would the Surge API be primarily for stateless surge? One other key controller to think about here is StatefulSet. During updates StatefulSet can respect a `maxUnavailable` field: https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/961-maxunavailable-for-statefulset. This effectively controls the velocity of the update. MaxSurge achieves the same thing, but with surge rather than unavailability.

The other complexity is around pod deletion ordering. For StatefulSet, this may matter. Rather than having a Surge API that reads the maxSurge parameter and delete pods, could the disruption controller just defer to the operator's standard pod update procedure? This may not work as well in the case where there is only a partial maintenance event (in this case the operator may want to optimize which pods need to be updated). The benefit is that you don't need to re-declare the maxSurge parameter on a Surge API, you can just have a passthrough of responsibility of which pods to delete and in which order to the controller/operator.

Author: We will probably scrap the Surge API  and start simpler. We can revisit it later if needed. I will update the KEP to reflect that.

Author: I have removed the maxSurge API.

- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

This KEP proposes adding a declarative API to manage a node maintenance. This API can be used to
implement additional capabilities around node draining.

## Motivation
The goal of this KEP is to analyze and improve node maintenance in Kubernetes.

Node maintenance is a request from a cluster administrator to remove all pods

> from a node, so that it can be disconnected from the cluster to perform a 

Reviewer 5: ```suggestion
from a node(s), so that it can be disconnected from the cluster to perform a 
```

   `.spec.drain=false`.-
3. Drain: gives an intent to drain all selected nodes by setting a `EvacuationRequest`
   condition with `Reason="NodeMaintenance"` on the node's pods. Configured with
   `.spec.cordon=true` and `.spec.drain=true`.
4. Drain Complete: all targeted pods have been drained from all the selected nodes. The nodes can
   be upgraded, restarted, or shut down. The configuration is still kept at `.spec.cordon=true` and
   `.spec.drain=true`.
5. Maintenance Complete: make the nodes schedulable again once the node maintenance is done.
   Set `.spec.cordon=false` and `.spec.drain=false` back again.

```golang

type NodeMaintenance struct {
    ...
    Spec NodeMaintenanceSpec
    Status NodeMaintenanceStatus
}

type NodeMaintenanceSpec struct {
    // +required
    NodeSelector *v1.NodeSelector
    // When set to true, cordons all selected nodes by making them unschedulable.
    Cordon  bool
    // When set to true, gives an intent to drain all selected nodes by setting
    // an EvacuationRequest condition on the node's pods.
    //
    // Drain cannot be set to true, unless Cordon is also set to true.

>     Drain bool

Reviewer 5: Instead of a series of bools, I'd suggest having a set of predefined states you expect. This will allow you to expand the API surface w/o adding more fields.

Author: I was planning to ask this in the API review, but yeah maybe it would be better to start with the more advanced approach first.

Author: I have added a `stage` field instead and added a section for supported stage transitions.

    Reason string
}

type NodeMaintenanceStatus struct {
    // Mapping of a node name to the maintenance status.
    // +optional

>     Nodes map[string]NodeMaintenanceNodeStatus

Reviewer 5: Usually APIs use just arrays. So I'd probably just move the name of the node (or even better NodeReference object), to NodeMaintananceNodeStatus.

Author: Ack. 

Although I am not sure if we need the reference since people will be usually selecting the nodes by names or labels.

Are you thinking of introducing new `NodeReference` type or just using `ObjectReference`?

Reviewer 5: > Are you thinking of introducing new `NodeReference` type or just using `ObjectReference`?

The general guideline is to have per-type references, so you'd create `NodeReference`

Author: I have moved the name into the `nodeRef`. We can add additional fields to the reference in the future if needed.

Author: and using arrays now

      how can an operator troubleshoot without logging into a master or worker node?
    - Mitigations: What can be done to stop the bleeding, especially for already
      running user workloads?
    - Diagnostics: What are the useful log messages and their required logging
      levels that could help debug the issue?
      Not required until feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.
-->

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

> ## Future Improvements

Reviewer 5: Any plans for statefulsets?

Author: I can mention it there. I would prefer not to do it in the first iteration. I think we need an option on the `NodeMaintenance` or/and on the `StatefulSet` to support this. The main problem is that it could be a disruptive action (e.g. StatefulSet with 1 pod) and we are trying to design the API to give more control to the application.

Reviewer: **After the first alpha**, a higher-level controller that owns a StatefulSet could manage scale out for the StatefulSet to account for the upcoming (or currently happening) maintenance.

There is no one-size-fits-all answer. For some apps, the right fix might be to add another temporary replica. For another app, it is to quiesce, snapshot and then delete the running Pod so that a replacement can come up in the same zone, avoiding the node that is tainted due to maintenance. In another situation, maybe there are always exactly 3&times;<i>n</i> replicas, and we use in-place vertical scaling&hellip;

Does this sound familiar? It's the [operator pattern](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/), right? My advice is to offer the NodeMaintenance API and teach higher-level abstractions - outside Kubernetes' core - to care about it. We can even offer an official CRD for older clusters.

Reviewer 6: FYI The use cases the users brought up to me are mostly statefulset related so far.

Author: @soltysh @sftim @dchen1107 

ReplicaSet, Deployment, StatefulSet are discussed in the https://github.com/kubernetes/enhancements/pull/4565 KEP now.
I have added empty sections to this KEP for DaemonSet and Static pods. And will discuss it there.

- when the KEP was retired or superseded
-->

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

## Future Improvements

### Disruption Controller: Eviction Protection and Observability

With the `.spec.MinAvailable` and `.spec.MaxUnavailable` options, PDB could unfortunately allow
surge pods to be evicted because `.status.CurrentHealthy` could become higher than
`.status.desiredHealthy`. This could disrupt ongoing migration of an application. To support
mission-critical applications, we might do one of the following:
- Consider pods with an `EvacuationInitiated` condition as disrupted and thus decreasing
  `.status.CurrentHealthy`.
- Or increase the `.status.desiredHealthy` by the number of pods with the `EvacuationInitiated`
  condition. This would keep the `.status.DisruptionsAllowed` value low enough not to disrupt the
  migration.

We can also count the evacuating pods in the status for observability, as this feature changes the
behavior of the PodDisruptionBudget status.

```golang
type PodDisruptionBudgetStatus struct {
    ...
    // total number of pods that are evacuating and expected to be terminated
    EvacuatingPods int32
    ...
}
```

> 

Reviewer: A future improvement could be to set a boolean condition on a workload management object (StatefulSet, ReplicaSet, Job etc) if any Pod managed by that object is matched by any non-expired NodeMaintenance.

_This feedback should not block merging the KEP as provisional._

Author: mentioning this in https://github.com/kubernetes/enhancements/pull/4565 in the `Follow-up Design Details for Kubernetes Workloads` section

You do not need a new KEP to move from beta to GA, for example. If
new details emerge that belong in the KEP, edit the KEP. Once a feature has become
"implemented", major changes should get new KEPs.

The canonical place for the latest set of instructions (and the likely source
of this file) is [here](/keps/NNNN-kep-template/README.md).

**Note:** Any PRs to move a KEP to `implementable`, or significant changes once
it is marked `implementable`, must be approved by each of the KEP approvers.
If none of those approvers are still appropriate, then changes to that list
should be approved by the remaining approvers and/or the owning SIG (or
SIG Architecture for cross-cutting KEPs).
-->
# KEP-4212: Declarative Node Maintenance

<!--
This is the title of your KEP. Keep it short, simple, and descriptive. A good
title can help communicate what the KEP is and should be considered as part of
any review.
-->

<!--
A table of contents is helpful for quickly jumping to sections of a KEP and for
highlighting any additional information provided beyond the standard KEP
template.

Ensure the TOC is wrapped with
  <code>&lt;!-- toc --&rt;&lt;!-- /toc --&rt;</code>

> tags, and then generate with `hack/update-toc.sh`.

Reviewer 7: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/enhancements/4213/pull-enhancements-verify/1754855493689413632#1:build-log.txt%3A911
> Table of content not up to date. Did you run 'make update-toc' ?

any required manual interventions. I want to have an ability to manually switch between the
maintenance phases (Planning, Cordon, Drain, Drain Complete, Maintenance Complete). I also want to
observe the node drain via the API and check on its progress. I also want to be able to discover
workloads that are blocking the node drain.

#### Story 2

As an application owner, I want to run single replica applications without disruptions and have the
ability to easily migrate the workload pods from one node to another.

#### Story 3

Cluster or node autoscalers that take on the role of `kubectl drain` want to signal the intent to
drain a node using the same API and provide a similar experience to the CLI counterpart.

#### Story 4

I want to be able to use a similar approach for general descheduling of pods that happens outside
of node maintenance.

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.

> -->

Reviewer: - If someone deletes a NodeMaintenance whilst the drain is ongoing, do we remove the EvacuationRequested condition on Pods already marked?

Reviewer: - If a Pod is marked `EvacuationInitiated` and the NodeMaintenance object(s) targeting that Pod are removed / postponed / whatever, what should happen?

(my opinion: **proceed**; evacuation has been initiated so we should continue it and expect a higher level abstraction to make a replacement Pod).

Author: Do you mean function-wise or name-wise?
- If the former, we need a way to go through the stages manually.
- If the later, I have tried to find a better name but could not decide on a better one. One of them is a phase, but they are meant as a cyclic event and I want the stage transitions to be fairly rigid and not to return back. So that there is a defined lifecycle of a single NodeMaintenance.

Author: I suppose in general yes. We do not know about the application behavior and it may not support cancellation (e.g. Pod is already terminating). Ultimately it should be up to the controller. It could react on the removal of the EvacuationRequest, but it doesn't have to..

Author: So I think it is better to always remove the `EvacuationRequest` condition, so that our intent is clear.

Author: added a mention for this

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

### Kubectl

`kubectl drain` command will be changed to create a NodeMaintenance object instead of marking the
node unschedulable. We will also change the implementation to skip applications that support

> workload migration. This will be detected by observing a `EvacuationRequest` condition on the pod

Reviewer: ```suggestion
workload migration. This will be detected by observing a `EvacuationRequested` condition on the pod
```
?

Author: I imagine this is a long standing request that is also reconciled and that can be removed. So I am not sure if the past tense is justified.

Reviewer: The past tense can fit even for ongoing situations.

Imagine the conditions on a building; we in the UK talk about some of our oldest buildings (castles) having been “slighted”, and some of our worst buildings being “condemned”. Many buildings that we cherish have been “listed”. Those conditions can last for hundreds of years!

Author: Nice comparison :) I have changed it to `EvacuationRequested`.

As a consequence the node maintenance controller will reconcile the node back to the old value.

### NodeMaintenance API

NodeMaintenance objects serve as an intent to remove or migrate pods from a set of nodes. We will
include Cordon and Drain toggles to support the following states/stages of the maintenance:
1. Planning: this is to let the users know that maintenance will be performed on a particular set
   of nodes in the future. Configured with `.spec.stage=Idle`.
   2. Cordon: stop accepting (scheduling) new pods. Configured with `.spec.stage=Cordon`.
3. Drain: gives an intent to drain all selected nodes by setting an `EvacuationRequest`
   condition with `Reason="NodeMaintenance"` on the node's pods. Configured with
   `.spec.stage=Drain`.
4. Drain Complete: all targeted pods have been drained from all the selected nodes. The nodes can
   be upgraded, restarted, or shut down. The configuration is still kept at `.spec.stage=Drain` and
   `Drained` condition is set to `"True"` on the node maintenance object.
5. Maintenance Complete: make the nodes schedulable again once the node maintenance is done.

>    Configured with `.spec.stage=Complete`.

Reviewer: `.spec.stage` sounds more like _status_ field to me.

```golang

// +enum
type NodeMaintenanceStage string

const (
// Idle does not interact with the cluster.
Idle NodeMaintenanceStage = "Idle"
// Cordon cordons all selected nodes by making them unschedulable.
Cordon NodeMaintenanceStage = "Cordon"
// Drain:
// 1. Cordons all selected nodes by making them unschedulable.
// 2. Gives an intent to drain all selected nodes by setting an EvacuationRequest condition on the
//    node's pods.
Drain NodeMaintenanceStage = "Drain"
// Complete:
// 1. Removes EvacuationRequest on all the pods targeted by this NodeMaintenance.
// 2. Uncordons all selected nodes by making them schedulable again, unless there is not another
//    maintenance in progress.
Complete NodeMaintenanceStage = "Complete"
)

type NodeMaintenance struct {
    ...
    Spec NodeMaintenanceSpec
    Status NodeMaintenanceStatus
}

type NodeMaintenanceSpec struct {
    // +required
    NodeSelector *v1.NodeSelector

>     // The order of the stages is Idle -> Cordon -> Drain -> Complete.

Reviewer: Post-alpha, we might want to support Idle → Cancelled and Idle → Cordon → Cancelled, to record that we never triggered the drain.

Author: I have to update the documentation here, but it should be similar to what can be found in the `Supported Stage Transitions` section.

A good idea: it would be good to record the passed stages in the status.

    //
    // - The Cordon or Drain stage can be skipped by setting the stage to Complete.
    // - The NodeMaintenance object is moved to the Complete stage on deletion unless the Idle stage has been set.
    //
    // The default value is Idle.

>     Stage NodeMaintenanceStage

Reviewer: Alternative (to be added later in the document): bring more details within Kubernetes' own purview.

The API defines a list (array) of stages; you can only add to the array and you can only add valid transitions.

Example transition chains:
- Upcoming → Cordon → Drain → Completed
- Upcoming → Cancelled
- Cordon → Drain → Cancelled
- Upcoming → Cordon → Drain → TerminateNode → Completed (_TerminateNode_ might be something we add for beta)

In YAML it might look like:
```yaml
…
  maintenanceStages:
    - stage: Upcoming
    - stage: Cordon
      maxStageDurationSeconds: 60
    - stage: Drain
      maxStageDurationSeconds: 1800
```

the node this affects are therefore expected to be mid-drain; the control plane adds a Completed stage after the duration limit times out.

This feels more extensible.

Author: I also had this idea in mind too, but I am still not sure. The stage is sort of like a status field but one that can be changed by the user or a controller mid-flight.

Maybe we can have an additional field like `maintenancePlan`? I have to think this through. Also to get it in-line with the additional features.

Reviewer: We list alternatives to tell people what we've ruled out; they go there despite _not_ being our favorite approach.

Author: I have introduced a drainPlan to the API. And also StageStatuses which mention which tracks the stages that were started and their timestamps.

### Finalizers and Deletion

When a stage is not `Idle`, `nodemaintenance.kubernetes.io/maintenance-completion` finalizer will be put on
the NodeMaintenance object to ensure uncordon and reset of EvacuationRequest conditions upon deletion.

When a deletion of the NodeMaintenance object is detected, its `.spec.stage` is set to `Complete`.
The finalizer is not removed until the `Complete` stage has been completed.

#### Cordon

When a `Cordon` or `Drain` stage is detected on the NodeMaintenance object, the controller
will set `.spec.Unschedulable` to `true` on all nodes that satisfy `.spec.nodeSelector`.

#### Uncordon (Complete)

When a `Complete` stage is detected on the NodeMaintenance object, the controller will:
- Set `.spec.Unschedulable` back to `false`  on all nodes that satisfy `.spec.nodeSelector`, unless
  there is no other maintenance in progress.
- Reset EvacuationRequest conditions on targeted pods (see Drain below for more details).

#### Drain

When a `Drain` stage is detected on the NodeMaintenance object, the
`EvacuationRequest` condition is set on selected pods. The condition should have

> `Reason="NodeMaintenance$NAME"` and `Message="$REASON""`, where `NAME` is the name of the
> NodeMaintenance object and `REASON` is equal to `.spec.reason`.

Reviewer: ```suggestion
`Reason="NodeMaintenance` and `Message="$REASON""`, where `REASON` is equal to `.spec.reason`.
```

the `reason` field of conditions is supposed to be opaquely machine-readable; let's keep that.

- For example, one day we might support a way to query Pods based whether they have a condition with a particular reason.
- if someone creates a NodeMaintenance named (eg) `Cancelled` this could confuse people/

Author: I agree. I have been discussing this already in https://github.com/kubernetes/enhancements/pull/4479#issuecomment-1933246115 and need to figure out something better.

Author: All these evacuation issues should be solved by the new Evacuation KEP https://github.com/kubernetes/enhancements/pull/4565

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

This KEP proposes adding a declarative API to manage a node maintenance. This API can be used to
implement additional capabilities around node draining.

## Motivation
The goal of this KEP is to analyze and improve node maintenance in Kubernetes.

Node maintenance is a request from a cluster administrator to remove all pods
from a node(s) so that it can be disconnected from the cluster to perform a
software upgrade (OS, Kubelet), hardware upgrade, or simply to remove the node
as it is no longer needed.

Kubernetes has existing support for this use case in the following way with `kubectl drain`:
1. There are running pods on node A, some of which are protected with PodDisruptionBudgets (PDB).

> 2. Set the node `Unschedulable` (cordon) to prevent new pods from being scheduled there.

Reviewer 8: Sharing experience here (we have been doing automatic node maintenance for more than 3y, with at least 3k nodes per month in our infra):
In our first version of the automation we were using `cordon` flag and it quickly became a problem. Since users can also play with that flag we started to face conflict:
- difficulty to understand if the flag was set or not by our automation
- users changing the value of the flag (uncordon) without understanding that they were interfering with the automation

The solution was to finally use a different taint key with effect `doNotSchedule`. It turned out to be a good choice especially because you can play with multiple taint key to reflect the stage of your automation. (drainCandidate - draining - drained)

Reviewer: In that story, what thing removed Pods from nodes where Pods were already executing?

Author: tolerations also have issues, the pods can tolerate your key or they can also tolerate everything

The advantage of cordon is that it is an existing primitive that is supported by kubectl commands. We can solve the interference problem by reconciling the field and making sure the correct value is always on the node. NodeMaintenance is a higher abstraction object, so if the users are not happy about this they should contact an admin to remove the NodeMaintenance from their node.

Nevertheless you can still go behind the scheduler's back and just set the nodeName directly.

Reviewer 8: > In that story, what thing removed Pods from nodes

We have a dedicated controller that handle the eviction. This controller is the one dealing with thousand of nodes "drain". I put "drain" between quote because in fact the controller is not calling "kubectl drain". It does the eviction of each individual pod, and we have a lot of feature associated to control pod eviction. Some of these feature are implemented over the PDB. The PDB spec is controlled by another controller that is able to integrate signal external to kubernetes to pilot the budget. We are also able to do cross-cluster validation. ... 
If you want to know more about that I invite you to look at https://datadogon.datadoghq.com/episodes/datadog-on-kubernetes-node-management/ ... of course you can skip the intro that is meant to introduce concepts that you must already know.

> tolerations also have issues, the pods can tolerate your key or they can also tolerate everything

true, but in practice, the toleration we are talking about is not even known by application users. It is a matter of ownership.

> The advantage of cordon is that it is an existing primitive that is supported by kubectl commands

Yes this is also precisely the problem that I was pointing: existing and used, and so you will have conflict with existing scripts or users and your automation. Talking about automation we have on our side introduce a dedicated API where the user can just set a dedicated annotation that finally gives the exact same result. The controller interpret that annotation and apply the required logic (taint, drain, ...)

>  We can solve the interference problem by reconciling the field and making sure the correct value is always on the node

You can, we also had that same idea at some point. Till some users created their own controller to cordon/uncordon node during regular operation for their application. The result was a nice fight between the 2 controllers with cordon/uncordon loops on a given node (and of course associated stress on the api-server and other controllers reacting to the updates).

Reviewer: > difficulty to understand if the flag was set or not by our automation

We have Server Side Apply which can help, and we (Kubernetes) could teach a tool like `kubectl` to warn if trouncing an existing field manager for `.spec.unschedulable`

Author: I understand that is possible to automate this in your case, but that is not a case for all k8s users today. The mission of these two KEPs is to have an extensible approach that everyone can use and hook into if needed.

I suppose, both approaches could also be used to ensure that the least amount of pods are scheduled to the targeted nodes during the conflict. We can also notify the admins via events if the conflict occurs, so that only a single node maintenance approach is used.

Author: updated relevant sections to solve this problem:
- https://github.com/kubernetes/enhancements/blob/a15b78d3b4f8bc1b4a4e5b1508e4ad733e64c4fd/keps/sig-apps/4212-declarative-node-maintenance/README.md#kubectl
- https://github.com/kubernetes/enhancements/blob/a15b78d3b4f8bc1b4a4e5b1508e4ad733e64c4fd/keps/sig-apps/4212-declarative-node-maintenance/README.md#cordon

@dbenque please open a new thread if needed

   (optionally by priority) by the cluster admin. This does not take into account 
   `.spec.terminationGracePeriodSeconds` of each pod and may cause premature termination of
   the application. This has been discussed in issue [kubernetes/kubernetes#75482](https://github.com/kubernetes/kubernetes/issues/75482)
   and in issue [kubernetes-sigs/cluster-api#6158](https://github.com/kubernetes-sigs/cluster-api/issues/6158).
4. There are cases during a node shutdown, when data corruption can occur due to premature node
   shutdown. It would be great if applications could perform data migration and synchronization of 
   cached writes to the underlying storage before the pod deletion occurs. This is not easy to
   quantify even with pod's `.spec.shutdownGracePeriod`, as the time depends on the size of the data
   and the speed of the storage. This has been discussed in issue [kubernetes/kubernetes#116618](https://github.com/kubernetes/kubernetes/issues/116618)
   and in issue [kubernetes/kubernetes#115148](https://github.com/kubernetes/kubernetes/issues/115148).
5. There is not enough metadata about why the node drain was requested. This has been discussed in

>    issue [kubernetes/kubernetes#30586](https://github.com/kubernetes/kubernetes/issues/30586).

Reviewer 9: NMO includes a [reason](https://github.com/medik8s/node-maintenance-operator/blob/main/api/v1beta1/nodemaintenance_types.go#L51-L53) for the maintenance procedure.

Furthermore, [NMO's status](https://github.com/medik8s/node-maintenance-operator/blob/main/api/v1beta1/nodemaintenance_types.go#L57-L85) adds some in-progress and post-progress information on the maintenance status. The operator also [emits events](https://github.com/medik8s/node-maintenance-operator/blob/main/pkg/utils/events.go) about the maintenance process.

Approaches and workarounds used by other projects to deal with these shortcomings:
- https://github.com/medik8s/node-maintenance-operator uses a declarative approach that tries to
  mimic `kubectl drain` (and uses kubectl implementation under the hood).
- https://github.com/kubereboot/kured performs automatic node reboots and relies on `kubectl drain`
  implementation to achieve that.
- https://github.com/strimzi/drain-cleaner prevents Kafka or ZooKeeper pods from being drained
  until they are fully synchronized. Implemented by intercepting eviction requests with a
  validating admission webhook. The synchronization is also protected by a PDB with the
  `.spec.maxUnavailable` field set to 0. See the experience reports for more information.
- https://github.com/kubevirt/kubevirt intercepts eviction requests with a validating admission
  webhook to block eviction and to start a virtual machine live migration from one node to another.
  Normally, the workload is also guarded by a PDB with the `.spec.minAvailable` field set to 1.
  During the migration the value is increased to 2.

> Experience Reports:

Reviewer 10: IMHO a lot of what is in this section is pretty anecdotal. It is from people/projects that were vocal but its unclear if what what they are describing is common.

OTOH what is extremely commonly used an notably absent here is how the three big CSPs (AWS, Azure, GCP) do node draining in their respective nodepool upgrade functionality of their managed k8s offerings.

Author: I have added the CSP use cases. However most of them use the Cluster Autoscaler.

Experience Reports are only part of the puzzle. There are a lot of relevant upstream issues and customer issue that experience these common problems. And blogs written by the users help to point that out.

@alvaroaleman please let me know if you have other sources you would like to add?

  application.

To sum up. Some projects solve this by introducing validating admission webhooks. This has a couple
of disadvantages. The webhooks are not easily discoverable by cluster admins. And they can block
evictions for other applications if they are misconfigured or misbehave. It is not intended for the
eviction API to be extensible in this way. The webhook approach is therefore not recommended.

As seen in the experience reports and GitHub issues, some admins solve their problems by simply
ignoring PDBs which can cause unnecessary disruptions or data loss. Some solve this by playing
with the application deployment, but have to understand that the application supports this.

### Goals
- Kubectl drain should not evict and disrupt applications with evacuation capability and instead
  politely ask them to migrate their pods to another node or to remove them by creation of
  NodeMaintenance object.
- Introduce a node maintenance controller that will help controllers like deployment controller
  to migrate their pods.
- Deployment controller should use `.spec.strategy.rollingUpdate.maxSurge` to evacuate its pods
  from a node that is under maintenance.

> ### Non-Goals

Reviewer 10: This section is titled `Non-Goals` and describes a list of things in detail which is very confusing. Can we reduce this to `Changes to how PDB status is calculated`, `Change behavior of DaemonSets`, `Prioritization of pods` etc?

Author: goals/non-goals have been updated

- The PDB controller should detect and account for applications with evacuation capability when
  calculating PDB status.

> - Introduce a field that could include non-critical daemon set pods

Reviewer 10: I find it very dubious to make any assumptions about when to terminate a daemonset pod based on its priority class. The point of a daemonset is to make sure that a given pod runs on a set of nodes for the lifetime of the node.  It is also not clear if not "gracefully shutting down daemon set pods" is a common problem in the first place. 

The fact that people reported this as a problem doesn't mean that this being a problem is the default case, just that it happens at times. People that are fine with what happens currently are not going to open a GitHub issue about that, so by looking just at GitHub issues we get a very skewed picture.

  pods of certain priorities (see point above). We could also introduce drain profiles based on
  these lists. The cluster admin could then choose or create such a profile based on his/her needs.
  The logic for processing the decision list would be contained in the node maintenance controller,
  which would set an intent to selected pods to shut down via the EvacuationRequested
  condition.
- Introduce a node maintenance period, nodeDrainTimeout (similar to [cluster-api](https://cluster-api.sigs.k8s.io/developer/architecture/controllers/control-plane)
  nodeDrainTimeout) or a TTL optional field as an upper bound on the duration of node maintenance.
  Then the node maintenance would be garbage collected and the node made schedulable again.

## Proposal
Most of these issues stem from missing a standardized way of detecting a start of the node drain.
This KEP proposes the introduction of a NodeMaintenance object that would signal an intent to
gracefully remove pods from given nodes. The application pods should then signal back that the pods
are being removed or migrated from the node. The implementation should also utilize existing node's

> `.spec.unschedulable` field, which prevents new pods from being scheduled on such a node.

Reviewer 8: https://github.com/kubernetes/enhancements/pull/4213/files#r1537798784

This field, can also be manipulated by users... having a dedicated taint with NoSchedule effect would remove that "concurrency/ownership" problem and open for even more control.

Reviewer 8: Let's close that thread since the same discussion is happening in the other comment.

We will focus primarily on `kubectl drain` as a consumer of the NodeMaintenance API, but it can
also be used by other drain implementations (e.g. node autoscalers) or manually. We will first
introduce the API and then later modify the behavior of the Kubernetes system to fix all the node
drain issues we mentioned earlier.

To support workload migration, a new controller should be introduced to observe the NodeMaintenance
objects and then mark pods for migration or removal with a condition. The pods would be selected
according to the node at first (`nodeSelector`), but the selection mechanism can be extended later.
Controllers can then implement the migration. The advantage of this approach is that controllers do

> not have to be aware of the NodeMaintenance object (no RBAC changes required). They only have to

Reviewer 9: >The advantage of this approach is that controllers do
not have to be aware of the NodeMaintenance object (no RBAC changes required)

Can you please elaborate why the controllers won't be aware of the NodeMaintenance object? Shouldn't they act upon changes (creation/deletion/update) to this object? Should they be aware of node changes (maintenanced node's status is Not Ready)?

Reviewer: Doing it this way means we can add the NodeMaintenance API without needing to change the whole ecosystem ahead of day 1 (which is handy!)

Reviewer 9: I understand the consequences, but I am not sure how it will be implemented in a way to achieve this goal.
How can the NodeMaintenance controller would not be aware of the NodeMaintenance object and still be able to act upon its creation/update/deletion?

Reviewer: Ah, try “controllers (apart from the NodeMaintenance controller)  do not have to be aware of the NodeMaintenance object kind”

The controller for NodeMaintenance does need to be coded; other controllers don't need updates.

Reviewer 9: Oh yes, thanks for elaborating. It clarified my thoughts.

Author: The RBAC argument does not apply anymore. Currently some controllers need to be only aware of the Evacuation API (https://github.com/kubernetes/enhancements/pull/4565) and some of both the Evacuation API and the NodeMaintenance API (TBD the daemonset controller in this KEP)

This might be a good place to talk about core concepts and how they relate.
-->

#### Cancelled Node Maintenance and Impact on Applications

When node maintenance is cancelled (reaches `Complete` stage without all of its pods terminating),
the node maintenance controller will remove `EvacuationRequested` conditions on the targeted pods.

1. Pods that have not yet initiated evacuation (no `EvacuationInitiated` condition) will
   continue to run unchanged.
2. Pods that have initiated evacuation have an option for their controller to either cancel the
   evacuation and remove the `EvacuationInitiated` condition, or keep the condition and finish the
   evacuation.

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

### Kubectl

`kubectl drain` command will be changed to create a NodeMaintenance object instead of marking the

> node unschedulable. We will also change the implementation to skip applications that support

Reviewer 9: I like this idea!

Author: The evacuation and node maintenance process has changed quite a bit since then.

workload migration. This will be detected by observing a `EvacuationRequested` condition on the pod

> and the subsequent appearance of `EvacuationInitiated` condition within a reasonable timeframe (3m).

Reviewer 9: How the 3m timeframe was selected?

Reviewer 11: Additionally, is there any way to avoid this wait-and-fallback behaviour for everything else, e.g. StatefulSets? Adding ~3m delay to drains/maintenance would be frustrating. Perhaps a way of a controller opting-in to say that it supports this evacuation requested api or not, and immediately failing back to previous logic if not?

Reviewer 11: Nevermind - it seems that the design has progressed significantly in https://github.com/kubernetes/enhancements/pull/4565#issuecomment-2045722703 towards dedicated resources with evacuators now advertising which pods they are able to evacuate.

Author: yup, there is no initial timeout anymore and the evacuation should be as responsive as the eviction

NodeMaintenance objects serve as an intent to remove or migrate pods from a set of nodes. We will
include Cordon and Drain toggles to support the following states/stages of the maintenance:
1. Planning: this is to let the users know that maintenance will be performed on a particular set
   of nodes in the future. Configured with `.spec.stage=Idle`.
   2. Cordon: stop accepting (scheduling) new pods. Configured with `.spec.stage=Cordon`.
3. Drain: gives an intent to drain all selected nodes by setting an `EvacuationRequested`
   condition with `Reason="NodeMaintenance"` on the node's pods. Configured with
   `.spec.stage=Drain`.
4. Drain Complete: all targeted pods have been drained from all the selected nodes. The nodes can
   be upgraded, restarted, or shut down. The configuration is still kept at `.spec.stage=Drain` and
   `Drained` condition is set to `"True"` on the node maintenance object.
5. Maintenance Complete: make the nodes schedulable again once the node maintenance is done.
   Configured with `.spec.stage=Complete`.

```golang

// +enum
type NodeMaintenanceStage string

const (
// Idle does not interact with the cluster.
Idle NodeMaintenanceStage = "Idle"

> // Cordon cordons all selected nodes by making them unschedulable.

Reviewer: > How can we enforce that there won't be two CRs that include the same node at the same time?

It's actually useful to be able to declare overlapping maintenance.

For example:
- something wants to be able to drain (evacuate) an entire datacenter
- during that overall work, a node fails, and a separate controller wants to schedule immediate maintenance on that one affected node

You can already `kubectl drain` a node covered by Medik8s; it's good if you still can after we make the change this KEP is calling for.

I can see that there'd be snags, too, but I think there's overall more value in allowing multiple requests. This aligns better with the real world: outside of Kubernetes, you can schedule a server to be shut down at the end of the week, and you can also tell someone to reboot it right now.

> // Cordon cordons all selected nodes by making them unschedulable.

Reviewer 9: I assume that by using the multiple nodes option we can still get a good insight into my above issue (one pod "blocks" the CR  change of stage) by meaningful information that would be stored in the CR status.

> // Cordon cordons all selected nodes by making them unschedulable.

Author: I think yes, it should be less of a problem if the observability of the node maintenance and evacuation processes is detailed/descriptive.

Also, there is nothing preventing users from creating NodeMaintenances node by node as before.

will set `.spec.Unschedulable` to `true` on all nodes that satisfy `.spec.nodeSelector`.

#### Uncordon (Complete)

When a `Complete` stage is detected on the NodeMaintenance object, the controller will:
- Set `.spec.Unschedulable` back to `false`  on all nodes that satisfy `.spec.nodeSelector`, unless
  there is no other maintenance in progress.
- Reset EvacuationRequested conditions on targeted pods (see Drain below for more details).

#### Drain

When a `Drain` stage is detected on the NodeMaintenance object, the
`EvacuationRequested` condition is set on selected pods. The condition should have
`Reason="NodeMaintenance$NAME"` and `Message="$REASON""`, where `NAME` is the name of the
NodeMaintenance object and `REASON` is equal to `.spec.reason`.
The pods would be selected according to the node (`.spec.nodeSelector`) and a subset of the default
kubectl [drain filters](https://github.com/kubernetes/kubernetes/blob/56cc5e77a10ba156694309d9b6159d4cd42598e1/staging/src/k8s.io/kubectl/pkg/drain/filters.go#L153-L162).

Used drain filters:
- `daemonSetFilter`, skips daemon sets to keep critical workloads alive.

> - `mirrorPodFilter`, skips static mirror pods.

Author: I have reworked the drain process. @razo7 PTAL. I would like this API to have as few toggles as possible to make its usage standardized and easy to use.

other maintenance in progress. The condition will only be removed if the reason of the condition
is `NodeMaintenance$NAME`, where NAME is the name of the NodeMaintenance. If the reason has a
different value, then it is owned by another controller (e.g. descheduler) or by a different
NodeMaintenance object and we should keep the condition.

The controller can show progress by reconciling:
- `.status.nodes[0].PodsPendingEvacuation`, to show how many pods remain to be removed from
  the first node.
- `.status.nodes[0].PodsEvacuating`, to show how many pods have been accepted for the
  evacuation from the first node. These are the pods that have the `EvacuationInitiated`
  condition set to `True`.
- To keep track of the entire maintenance the controller will reconcile a `Drained` condition and
  set it to true if all pods pending evacuation/termination have terminated on all target nodes
  when drain is requested by the maintenance object.
- NodeMaintenance condition or annotation can be set on the node object to advertise the current
  phase of the maintenance.

#### Supported Stage Transitions

The following transitions should be validated by the API server.

- Idle -> _Deletion_

>   - Planning a maintenance in the future and canceling/deleting it without any consequence.

Reviewer: Tracking planned maintenance allows an autoscaling controller to create a new node ahead of the maintenance starting, preserving cluster capacity.

It's a capacity surge model rather than a repair model.

>   - Planning a maintenance in the future and canceling/deleting it without any consequence.

Reviewer 9: I see the value of the idle stage, but what would initiate moving from _Idle_ to _Cordon_. I didn't understand how and when it will change.

>   - Planning a maintenance in the future and canceling/deleting it without any consequence.

Author: It is an API, so either manually by the cluster admin or by a controller. Note that the user can create the NodeMaintenance with `stage: Drain` directly. This is just provides more options.

  - The complete stage will always be run even without specifying it.
- (Idle) -> (Cordon) -> Drain -> (Complete) -> _Deletion_.
  - Make a set of nodes unschedulable, drain them, and then make them schedulable again.
  - Cordon and Complete stages will always be run, even without specifying them.
- (Idle) -> Complete -> _Deletion_.
  -  Make a set of nodes schedulable.

### Deployment and ReplicaSet Controllers

The replica set controller will watch its pods and count the number of pods it observes with a
`EvacuationRequested` condition. It will then store this count in `.status.ReplicasToEvacuate`

The deployment controller will watch its ReplicaSets and react when it observes positive number of
pods in `.status.ReplicasToEvacuate`. If the owning object of the targeted pods is a Deployment with
a positive `.spec.strategy.rollingUpdate.maxSurge` value, the controller will create surge pods by
scaling up the ReplicaSet. The new pods will not be scheduled on the maintained node because the
`.spec.unschedulable` field would be set to true on that node. As soon as the surge pods become
available, the deployment controller will scale down the ReplicaSet. The replica set controller

> will then in turn delete the pods with the `EvacuationRequested` condition.

Reviewer 10: Could you please also spell out what happens if `maxSurge` is `0` and when the deployment uses the `Recreate` strategy?

Author: The Deployment controller has been moved to the Evacuation API KEP https://github.com/kubernetes/enhancements/pull/4565. PTAL

For completeness, the deployment controller will also track the total number of targeted pods of
all its ReplicaSets under its `.status.ReplicasToEvacuate`.

If the node maintenance prematurely ends before the surge process has a chance to complete, the
deployment controller will scale down the ReplicaSet which will then remove the extra pods that
were created during the surge.

```golang
type ReplicaSetStatus struct {
    ...
    ReplicasToEvacuate int32
    ...
}
```

```golang
type DeploymentStatus struct {
    ...
    ReplicasToEvacuate int32
    ...
}
```

To support providing a response to the drain process that the evacuation has begun. The deployment

> controller will annotate all replica sets that support the evacuation with the annotation

Reviewer 10: What does it mean `to support the evacution`? And why use an annotation and not a proper spec field?

Author: The evacuation process has been changed. PTAL at https://github.com/kubernetes/enhancements/pull/4565

`deployment.kubernetes.io/evacuation-ready`. For now this will apply to replica sets of deployments
with `.spec.strategy.rollingUpdate.maxSurge`. When this annotation is present, the replication
controller will respond to all evacuation requests by setting the `EvacuationInitiated` condition
to all of its pods with the `EvacuationRequested` condition.

### Node Drain Process: Before and After

The following diagrams describe how the node drain process will change in respective to each
component.

> Current state of node drain:

Reviewer 8: The diagram above is over simplified.
A drain consumes the eviction api for all the pods of the nodes, and not the Delete. This is a huge difference, because the eviction API creates the an eviction subresource. This creation lead to an admission-webhook call that allows platform admins to do a LOT that is going to be skipped with the solution. In our company we are doing multiple checks at eviction admission that will not be possible anymore with the solution. 
That would be a problem, for any user that is gating the pod eviction at this level to:
- validate that there is no global context that prevent that eviction. (During incident it could be convenient to block eviction by CA (repacking) or by other automation), (During operation freeze it could be interesting to block any user eviction request)...
- perform last synchronous check: if you application is distributed over multiple clusters for resiliency purpose, you don't want to create disruption/eviction in all the clusters at the same time.

To summarise, the loss of the admission-webhook on evictions could have a huge impact on operation by removing a convenient point of control. The proposed solution should come with similar capability. If not we are going to solve a problem but we are going to create another one.

Reviewer: Thanks for the review, you raise a good point here.

I actually think people might want a node maintenance controller to attempt graceful (eviction subresource) draining, and then optionally switch to forcible Pod removal.

For example, if the data center is scheduled to get powered down Tuesday, blocking the API-driver eviction with a webhook response or ValidatingAdmissionPolicy isn't going to keep the electricity supply going.

Author: Thank you for this important feedback. The admission webhooks intercepting the evictions are not supported in Kubernetes today. This introduces a lot of undefined behaviour that cannot be reasoned about by other components using the eviction API.

> To summarise, the loss of the admission-webhook on evictions could have a huge impact on operation by removing a convenient point of control. The proposed solution should come with similar capability. If not we are going to solve a problem but we are going to create another one.

This KEP and the Evacuation KEP https://github.com/kubernetes/enhancements/pull/4565 are trying to solve this problem. That is, you can subscribe to the evacuations and become the evacuator who then decides how the evacuation should proceed and wait for the global context and do before deletion checks.

The Evacuation object itself can be used as a lock to synchronize the evacuator ownership. It does not discuss the evacuator priority though, but I will try to outline it there. 

Even though the webhook approach is convenient, it is very opaque and hard to debug, so we try to go with a transparent approach.

@dbenque 
- Could you please open an issue with this feature requres in kubernetes/kubernetes so we can reference it in the KEP? 
- What mechanism do you think would be the best for handing over the ownership between the different evacuators?

Reviewer 8: @atiratree can you please point me at the documentation or code that justifies the following:
> The admission webhooks intercepting the evictions are not supported in Kubernetes today

I am a bit concerned by this statement because I have been using such an admission on eviction for more than 2y now and it works fine and it gives full satisfaction.

I will have a look to KEP https://github.com/kubernetes/enhancements/pull/4565 thanks for the link.

> it is very opaque and hard to debug, so we try to go with a transparent approach.

Not sure what you mean by "opaque", admission is a well known pattern in kubernetes. Sure it can be a double edge sword, and this is why as platform owner we try not to let users define/play with it... but for a platform owner it is a very powerful and convenient tool.
If we find a better way that offer the same capability (level of control) then that should be fine.

Reviewer: (I'm not sure about unsupported, but both the v1 Pod API and admission control through validating webhooks are generally available features)

Author: > I am a bit concerned by this statement because I have been using such an admission on eviction for more than 2y now and it works fine and it gives full satisfaction.

Admission webhooks are supported, but not everything what you do with them is. For example, if you fail all pod creations, you cannot be dissatisfied that the ReplicaSet does not create the pods and thus failing the conformance. The same goes for eviction and all the places that depend on it.

> Not sure what you mean by "opaque", admission is a well known pattern in kubernetes. Sure it can be a double edge sword, and this is why as platform owner we try not to let users define/play with it... but for a platform owner it is a very powerful and convenient tool.

It is opaque in the sense that it is not discoverable and observable by other actors who have not deployed or developed the admission plugins.

@dbenque I have added a support for multiple evacuators to the Evacuation KEP https://github.com/kubernetes/enhancements/pull/4565#issuecomment-2045722703 . It also tackles the issues of discoverability and observability. Please let me know if it solves your use case.

![drain-before](https://raw.githubusercontent.com/kubernetes/website/f2ef324ac22e5d9378f2824af463777182817ca6/static/images/docs/kubectl_drain.svg)

Proposed node drain:

> ![drain-after](node-maintenance.svg)

Reviewer 12: Hi, I come across this proposal and it looks interesting!

From this diagram, it shows that `kubectl drain NODE` creates the `NodeMaintenance` object with `.spec.stage=Drain`. What about `idle` stage? Who would create the object in `idle` to signal the start of planning?

Author: Today, I think we should discard the whole kubectl drain flow and move directly into the declarative realm.

Author: @wangzhen127 The KEP is using the declarative format now and discarding the kubectl drain use case. PTAL

  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

> This KEP proposes adding a declarative API to manage a node maintenance. This API can be used to
> implement additional capabilities around node draining.

Reviewer 13: ```suggestion
This KEP proposes adding a declarative API to manage node maintenance. This API can be used to
implement additional capabilities around node draining.
```

> ## Motivation

Reviewer 13: ```suggestion
## Motivation

```

> The goal of this KEP is to analyze and improve node maintenance in Kubernetes.
> 
> Node maintenance is a request from a cluster administrator to remove all pods
> from a node(s) so that it can be disconnected from the cluster to perform a
> software upgrade (OS, Kubelet), hardware upgrade, or simply to remove the node
> as it is no longer needed.

Reviewer 13: ```suggestion
The goal of this KEP is to analyze and improve node maintenance in Kubernetes.

Node maintenance is a request from a cluster administrator to remove all pods
from a node(s) so that it can be disconnected from the cluster to perform a
software upgrade (OS, kubelet, etc.), hardware or firmware upgrade, or simply
to remove the node as it is no longer needed.
```

Kubernetes has existing support for this use case in the following way with `kubectl drain`:
1. There are running pods on node A, some of which are protected with PodDisruptionBudgets (PDB).
2. Set the node `Unschedulable` (cordon) to prevent new pods from being scheduled there.
3. Evict (default behavior) pods from node A by using the eviction API (see [kubectl drain worklflow](https://raw.githubusercontent.com/kubernetes/website/f2ef324ac22e5d9378f2824af463777182817ca6/static/images/docs/kubectl_drain.svg)).

> 4. Proceed with the maintenance and shut down the node.

Reviewer 13: ```suggestion
4. Proceed with the maintenance and shut down (or restart) the node.
```

> 5. Kubelet can try to delay the shutdown to allow the remaining pods to terminate gracefully

Reviewer 13: ```suggestion
5. Kubelet will try to delay the shutdown to allow the remaining pods to terminate gracefully
```

Reviewer: Is that true @kwilczynski ? What's your thinking around using “will” here?

Reviewer 13: @sftim, capability vs possibility. Can it delay the shutdown? Yes. Will it do it? Yes. Provided that the feature is enabled (should be). But, I might be wrong about this. Thoughts?

Reviewer: We can't assume that the kubelet is able to delay a shutdown (might be a node which doesn't offer that option).

We could write
```suggestion
5. On nodes that support it, the kubelet may detect the imminent shutdown and then trigger graceful
   shutdown handling:
   * delay the shutdown pending graceful termination of remaining pods
   * terminate remaining pods in reverse priority order
   
   Only some nodes support this graceful shutdown handling.
```

Reviewer 13: We could.

To add more colour:

However, the current implementation of Graceful Node Shutdown relies on underlying node support, assuming that the feature is enabled and the platform supports it. We'd fail if we could not inhibit the shutdown. There will be no graceful shutdown of any sort if we cannot set an inhibitor for whatever reason.

As such, what about this the following:

> On platforms and nodes that support it, the kubelet will try to detect the imminent shutdown and then
> attempt to perform a graceful node shutdown.

Author: Thanks for the suggestions, updated

> 5. Kubelet can try to delay the shutdown to allow the remaining pods to terminate gracefully

Reviewer 13: ```suggestion
5. kubelet can try to delay the shutdown to allow the remaining pods to terminate gracefully
```

> 5. Kubelet can try to delay the shutdown to allow the remaining pods to terminate gracefully

Reviewer 13: Update per comment: https://github.com/kubernetes/enhancements/pull/4213/files#r1599470547

```suggestion
5. On platforms and nodes that support it, the kubelet will try to detect the imminent shutdown and then
attempt to perform a graceful node shutdown.
```

   ([graceful-node-shutdown](https://kubernetes.io/docs/concepts/architecture/nodes/#graceful-node-shutdown)). 

>    The Kubelet also takes pod priority into account ([pod-priority-graceful-node-shutdown](https://kubernetes.io/docs/concepts/architecture/nodes/#pod-priority-graceful-node-shutdown))

Reviewer 13: ```suggestion
   The kubelet also takes pod priority into account ([pod-priority-graceful-node-shutdown](https://kubernetes.io/docs/concepts/architecture/nodes/#pod-priority-graceful-node-shutdown))
```

> The main problem is that the current approach tries to solve this in an application agnostic way
> and just tries to get rid of all the pods on the node. Since this approach cannot be applied

Reviewer 13: ```suggestion
The main problem is that the current approach tries to solve this in an application agnostic way
and will attempt to remove all the pods currently running on the node. Since this approach
cannot be applied
```

   to edit the PDB to account for the additional pod that needs to be spun to move the workload
   from one node to another. This has been discussed in issue [kubernetes/kubernetes#66811](https://github.com/kubernetes/kubernetes/issues/66811)
   and in issue [kubernetes/kubernetes#114877](https://github.com/kubernetes/kubernetes/issues/114877).
2. Similar to the first point, it is difficult to use PDBs for applications that can have a variable
   number of pods; for example applications with a configured horizontal pod autoscaler (HPA). These
   applications cannot be disrupted during a low load when they have only pod. However, it is
   possible to disrupt the pods during a high load without experiencing application downtime. If
   the minimum number of pods is 1, PDBs cannot be used without blocking the node drain. This has
   been discussed in issue [kubernetes/kubernetes#93476](https://github.com/kubernetes/kubernetes/issues/93476).
3. Graceful deletion of DaemonSet pods is currently only supported as part of (Linux) graceful node
   shutdown. The length of the shutdown is again not application specific and is set cluster-wide

>    (optionally by priority) by the cluster admin. This does not take into account 

Reviewer 13: The total time for which the shutdown will be inhibited takes this part of a given pod specification into account—it's the sum of each pod's provided value. See:

- [kubernetes/kubernetes/pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go)

For the concrete implementation.

Author: I see, so it can make the value lower, but the upper bound for the termination still holds. I have updated this point to reflect that.

4. There are cases during a node shutdown, when data corruption can occur due to premature node
   shutdown. It would be great if applications could perform data migration and synchronization of 
   cached writes to the underlying storage before the pod deletion occurs. This is not easy to
   quantify even with pod's `.spec.shutdownGracePeriod`, as the time depends on the size of the data
   and the speed of the storage. This has been discussed in issue [kubernetes/kubernetes#116618](https://github.com/kubernetes/kubernetes/issues/116618)
   and in issue [kubernetes/kubernetes#115148](https://github.com/kubernetes/kubernetes/issues/115148).
5. During the graceful node shutdown the kubelet terminates the pods in order of their priority.
   The daemonset controller runs its own scheduling logic and creates the pods again. This causes a
   race. Such pods should be removed and not recreated, but higher priority pods that have not yet
   been terminated should be recreated if they are missing. This has been discussed in issue
   [kubernetes/kubernetes#122912](https://github.com/kubernetes/kubernetes/issues/122912).

> 6. The graceful node shutdown feature is not always reliable. If dbus or kubelet is restarted

Reviewer 13: ```suggestion
6. The Graceful Node Shutdown feature is not always reliable. If dbus or kubelet is restarted
```

>    during the shutdown, pods may be ungracefully terminated and lead to application disruption and

Reviewer 13: Pods not only can be ungracefully terminated, but also sometimes the previously terminated workloads can be scheduled back on the node. kubelet will attempt to set the node status as such to allow scheduling on it to take place. That can also be undersired if not dangerous in some circumstances.

Author: ack, added

> 6. The graceful node shutdown feature is not always reliable. If dbus or kubelet is restarted
>    during the shutdown, pods may be ungracefully terminated and lead to application disruption and
>    data loss. This has been discussed in issues [kubernetes/kubernetes#122674](https://github.com/kubernetes/kubernetes/issues/122674),
>    [kubernetes/kubernetes#120613](https://github.com/kubernetes/kubernetes/issues/120613) and
>    [kubernetes/kubernetes#122674](https://github.com/kubernetes/kubernetes/issues/112443).

Reviewer 13: ```suggestion
6. The graceful node shutdown feature is not always reliable. If Dbus or kubelet is restarted
   during the shutdown, pods may be ungracefully terminated, leading to application disruption and
   data loss. This has been discussed in issues [kubernetes/kubernetes#122674](https://github.com/kubernetes/kubernetes/issues/122674),
   [kubernetes/kubernetes#120613](https://github.com/kubernetes/kubernetes/issues/120613) and
   [kubernetes/kubernetes#122674](https://github.com/kubernetes/kubernetes/issues/112443).
```

  implementation to achieve that.
- https://github.com/strimzi/drain-cleaner prevents Kafka or ZooKeeper pods from being drained
  until they are fully synchronized. Implemented by intercepting eviction requests with a
  validating admission webhook. The synchronization is also protected by a PDB with the
  `.spec.maxUnavailable` field set to 0. See the experience reports for more information.
- https://github.com/kubevirt/kubevirt intercepts eviction requests with a validating admission
  webhook to block eviction and to start a virtual machine live migration from one node to another.
  Normally, the workload is also guarded by a PDB with the `.spec.minAvailable` field set to 1.
  During the migration the value is increased to 2.
- https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler has an eviction process
  that takes inspiration from kubectl and build additional logic on top of it.
  See [Cluster Autoscaler](#cluster-autoscaler) for more details.
- https://github.com/kubernetes-sigs/karpenter taints the node during the node drain. It then
  attempts to evict all the pods on the node by calling the Eviction API. It prioritizes

>   non-critical pods and non-daemonset pods

Reviewer 13: ```suggestion
  non-critical pods and non-DaemonSet pods
```

- Pods and ReplicaSets/ReplicationControllers without owning controllers are blocking by default
  (the check can be modified with the `skip-nodes-with-custom-controller-pods` option).
- System pods (in the `kube-system` namespace) without a matching PDB are blocking by default
  (the check can be modified with the `skip-nodes-with-system-pods` option).
- Pods with `cluster-autoscaler.kubernetes.io/safe-to-evict: "false"` annotation are blocking.
- Pods with local storage are blocking unless they have a
  `cluster-autoscaler.kubernetes.io/safe-to-evict-local-volumes` annotation
  (the check can be modified with the `skip-nodes-with-local-storage` option, e.g., this check
  is skipped on [AKS](https://learn.microsoft.com/en-us/azure/aks/cluster-autoscaler?tabs=azure-cli#cluster-autoscaler-profile-settings)).
- Pods with PDBs that do not have `disruptionsAllowed` are blocking.

This can be enhanced with other rules and overrides.

It uses this logic to first check if all pods can be removed from a node. If not, it will report
those nodes. Then it will group all pods by a priority and evict them gradually from the lowest
to highest priority. This may include DaemonSet pods.

> ### Kubelet

Reviewer 13: ```suggestion
### kubelet
```

> Graceful node shutdown is a part of today's solution for node maintenance. Unfortunately it is not
> possible to rely solely on this feature as a go-to solution for graceful node and workload
> termination.

Reviewer 13: ```suggestion
Graceful Node Shutdown is a part of the current solution for node maintenance. Unfortunately, it is not
possible to rely solely on this feature as a go-to solution for graceful node and workload
termination.
```

> - The graceful node shutdown feature is not application specific and may disrupt workloads
>   prematurely and lead to data loss.

Reviewer 13: ```suggestion
- The Graceful Node Shutdown feature is not application-aware and may disrupt workloads
  prematurely and lead to data loss.
```

> - The kubelet implements the shutdown via the systemd inhibitor. If dbus or kubelet is restarted
>   during the node shutdown, the shutdown is not registered and pods are terminated ungracefully.
>   Also, new workloads can get scheduled on the node while the node is shutting down. Cluster admins
>   should therefore plan the maintenance in advance and ensure that pods are gracefully removed
>   before attempting to shut down or restart the machine.

Reviewer 13: ```suggestion
- The kubelet controls the shutdown process using Dbus and systemd, and can delay (but not entirely block) it 
  using the systemd inhibitor. However, if Dbus or kubelet is restarted during the node shutdown, the shutdown 
  might not be registered again, and pods might be terminated ungracefully. Also, new workloads can get 
  scheduled on the node while the node is shutting down. Cluster admins should, therefore, plan the 
  maintenance in advance and ensure that pods are gracefully removed before attempting to shut down or 
  restart the machine.
```

> - The kubelet has no way of detecting ongoing maintenance if the node is restarted in the meantime.

Reviewer 13: ```suggestion
- The kubelet has no way of reliably detecting ongoing maintenance if the node is restarted in the meantime.
```

> - The kubelet has no way of detecting ongoing maintenance if the node is restarted in the meantime.

Author: > Does this refer to the current state of affairs, or would this be in the context of this new KEP?

This is under the Motivation so it is the current state of affairs.

> If the node is restarted, the kubelet will be restarted along with it? Do we then expect the maintenance to continue afterwards?

Yes, we do, until the admin removes the NodeMaintenance.

> - Graceful termination of static pods during a shutdown is not possible today. It is also not

Reviewer 13: I believe Graceful Node Shutdown is not aware of whether a pod was started from a static template or via some other mechanism that scheduled it on the underlying node, so pods that were started as Static Pods will also be terminated in the same way as any other pods by the Graceful Node Shutdown manager.

Author: Not sure if I am reading the code correctly. But it seems to me that the shutdown manager obtains the pods as follows:

https://github.com/kubernetes/kubernetes/blob/a31030543c47aac36cf323b885cfb6d8b0a2435f/pkg/kubelet/kubelet.go#L923

which does not seem to include static/mirror pods.

> - Graceful termination of static pods during a shutdown is not possible today. It is also not
>   possible to keep them terminated after the machine has been restarted if the node is still under
>   maintenance. 

Reviewer 13: ```suggestion
- Graceful termination of static pods during a shutdown is not possible today. It is also currently not possible to 
  keep them from starting back up immediately after the machine has been restarted and kubelet has started 
  again, if the node is still under maintenance. 
```

> 

Reviewer 13: ```suggestion
```

### Motivation Summary

To sum up. Some applications solve the disruption problem by introducing validating admission
webhooks. This has some drawbacks. The webhooks are not easily discoverable by cluster admins. And
they can block evictions for other applications if they are misconfigured or misbehave. The
eviction API is not intended to be extensible in this way. The webhook approach is therefore not
recommended.

Some drainers solve the node drain by depending on the kubectl logic, or by extending/rewriting it
with additional rules and logic.

As seen in the experience reports and GitHub issues, some admins solve their problems by simply
ignoring PDBs which can cause unnecessary disruptions or data loss. Some solve this by playing
with the application deployment, but have to understand that the application supports this.

> Kubelet's graceful node shutdown feature is a best-effort solution for unplanned shutdowns, but

Reviewer 13: ```suggestion
kubelet's Graceful Node Shutdown feature is a best-effort solution for unplanned shutdowns, but
```

it is not sufficient to ensure application and data safety.

### Goals
- Introduce NodeMaintenance API.

> - Introduce a node maintenance controller that creates evacuations

Reviewer 13: ```suggestion
- Introduce a node maintenance controller that creates evacuations.
```

- Deprecate kubectl drain in favor of NodeMaintenance.

> - Implement NodeMaintenanceAwareKubelet feature to implement a lifecycle for Static pods during a

Reviewer 13: ```suggestion
- Implement NodeMaintenanceAwareKubelet feature to implement a lifecycle for Static Pods during a
```

  maintenance.
- Implement NodeMaintenanceAwareDaemonSet feature to prevent the scheduling of DaemonSet pods on
  nodes during a maintenance.

### Non-Goals
- Introduce a node maintenance period, nodeDrainTimeout (similar to [cluster-api](https://cluster-api.sigs.k8s.io/developer/architecture/controllers/control-plane)
  nodeDrainTimeout) or a TTL optional field as an upper bound on the duration of node maintenance.
  Then the node maintenance would be garbage collected and the node made schedulable again.

## Proposal

Most of these issues stem from the lack of a standardized way of detecting a start of the node
drain. This KEP proposes the introduction of a NodeMaintenance object that would signal an intent
to gracefully remove pods from given nodes. The intent will be implemented by the newly proposed
[Evacuation API KEP](https://github.com/kubernetes/enhancements/issues/4563), which ensures graceful pod removal or migration, an ability to measure
the progress and a fallback to eviction if progress is lost. The NodeMaintenance implementation

> should also utilize existing node's `.spec.unschedulable` field, which prevents new pods from being

Reviewer 13: Wouldn't this result in the same issues as what we have seen currently between the Graceful Node Shutdown manager and the DaemonSet controller? See:

- https://github.com/kubernetes/kubernetes/issues/122912

DaemonSet controller, through the nature of its design, will ignore most of the taints and Node statuses.

Author: Please look at the [DaemonSet Controller](https://github.com/kubernetes/enhancements/blob/a15b78d3b4f8bc1b4a4e5b1508e4ad733e64c4fd/keps/sig-apps/4212-declarative-node-maintenance/README.md#daemonset-controller) section. It will have to observe the NodeMaintenance API to adhere to the maintenance.

> Most of these issues stem from the lack of a standardized way of detecting a start of the node
> drain. This KEP proposes the introduction of a NodeMaintenance object that would signal an intent
> to gracefully remove pods from given nodes. The intent will be implemented by the newly proposed
> [Evacuation API KEP](https://github.com/kubernetes/enhancements/issues/4563), which ensures graceful pod removal or migration, an ability to measure
> the progress and a fallback to eviction if progress is lost. The NodeMaintenance implementation
> should also utilize existing node's `.spec.unschedulable` field, which prevents new pods from being
> scheduled on such a node.

Reviewer 13: ```suggestion
Most of these issues stem from the lack of a standardized way of detecting a start of the node
drain. This KEP proposes the introduction of a NodeMaintenance object that would signal an intent
to gracefully remove pods from given nodes. The intent will be implemented by the newly proposed
[Evacuation API KEP](https://github.com/kubernetes/enhancements/issues/4563), which ensures
graceful pod removal or migration, an ability to measure the progress and a fallback to eviction if
progress is lost. The NodeMaintenance implementation should also utilize existing node's
`.spec.unschedulable` field, which prevents new pods from being scheduled on such a node.
```

We will deprecate the `kubectl drain` as the main mechanism for draining nodes and drive the whole
process via a declarative API. This API can be used either manually or programmatically by other
drain implementations (e.g. cluster autoscalers). 

To support workload migration, a new controller should be introduced to observe the NodeMaintenance
objects and then select pods for evacuation. The pods should be selected by node (`nodeSelector`)
and the pods should be gradually evacuated according to the workload they are running.
Controllers can then implement the migration/termination either by reacting to the Evacuation API
or by reacting to the NodeMaintenance API if they need more details.

> 

Reviewer 13: ```suggestion
```

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

- This KEP depends on [Evacuation API KEP](https://github.com/kubernetes/enhancements/issues/4563).

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

### Kubectl

`kubectl drain`: as we can see in the [Motivation](#motivation) section, kubectl is heavily used
either manually or as a library by other projects. It is safer to keep the old behavior of this
command. However, we will deprecate it along with all the library functions. We can print a
deprecation warning when this command is used, and promote the NodeMaintenance. Additionally, pods

> that support evacuation and have `evacuation.coordination.k8s.io/priority_${EVACUATOR_CLASS}`

Reviewer: I'd use `-` not `_`.

    //   podType: "Default"
    // - podPriority: 2000001000 // system-node-critical priority class
    //   podType: "Default"
    // - podPriority: 2147483647 // maximum value
    //   podType: "Default"
    // - podPriority: 1000000000 // highest priority for user defined priority classes
    //   podType: "DaemonSet"
    // - podPriority: 2000000000 // system-cluster-critical priority class
    //   podType: "DaemonSet"
    // - podPriority: 2000001000 // system-node-critical priority class
    //   podType: "DaemonSet"
    // - podPriority: 2147483647 // maximum value
    //   podType: "DaemonSet"
    // - podPriority: 1000000000 // highest priority for user defined priority classes
    //   podType: "Static"
    // - podPriority: 2000000000 // system-cluster-critical priority class
    //   podType: "Static"
    // - podPriority: 2000001000 // system-node-critical priority class
    //   podType: "Static"
    // - podPriority: 2147483647 // maximum value
    //   podType: "Static"
    //
    // Duplicate entries are not allowed.
    // This field is immutable.
    DrainPlan []DrainPlanEntry

    // Reason for the maintenance.
    Reason string
}

const (

> // Default selects all pods except DaemonSet and Static pods.	

Reviewer 13: ```suggestion
// Default selects all pods except DaemonSet and static pods.	
```

Default PodType = "Default"
// DaemonSet selects DaemonSet pods.
DaemonSet PodType = "DaemonSet"
// Static selects static pods.
Static PodType = "Static"
)

type DrainPlanEntry struct {
    // PodSelector selects pods according to their labels.
	// This can help to select which pods of the same priority should be evacuated first.
    // +optional
    PodSelector *metav1.LabelSelector
    // PodPriority specifies a pod priority.
    // Pods with a priority less or equal to this value are selected.
    PodPriority int32
    // PodType selects pods according to the pod type:

>     // - Default selects all pods except DaemonSet and Static pods.	

Reviewer 13: ```suggestion
    // - Default selects all pods except DaemonSet and static pods.
```

> type DrainPlanEntry struct {
>     // PodSelector selects pods according to their labels.
> 	// This can help to select which pods of the same priority should be evacuated first.
>     // +optional
>     PodSelector *metav1.LabelSelector
>     // PodPriority specifies a pod priority.
>     // Pods with a priority less or equal to this value are selected.
>     PodPriority int32
>     // PodType selects pods according to the pod type:
>     // - Default selects all pods except DaemonSet and Static pods.	
>     // - DaemonSet selects DaemonSet pods.
>     // - Static selects static pods.
>     PodType PodType
> }

Reviewer 13: ```suggestion
type DrainPlanEntry struct {
    // PodSelector selects pods according to their labels.
    // This can help to select which pods of the same priority should be evacuated first.
    // +optional
    PodSelector *metav1.LabelSelector
    // PodPriority specifies a pod priority.
    // Pods with a priority less or equal to this value are selected.
    PodPriority int32
    // PodType selects pods according to the pod type:
    // - Default selects all pods except DaemonSet and Static pods.	
    // - DaemonSet selects DaemonSet pods.
    // - Static selects static pods.
    PodType PodType
}
```

  progressDeadlineSeconds: 1800
  evacuators:
    - evacuatorClass: deployment.apps.k8s.io
      priority: 10000
      role: controller
```

The node maintenance controller requests the removal of a pod from a node by the presence of the
Evacuation. Setting `progressDeadlineSeconds` to  1800 (30m) should give potential evacuators
enough time to recover from a disruption and continue with the graceful evacuation. If the
evacuators are unable to evacuate the pod, or if there are no evacuators, the evacuation controller
will attempt to evict these pods, until they are deleted.

The only job of the node maintenance controller is to make sure that the Evacuation object exist
and has the `evacuation.coordination.k8s.io/instigator_nodemaintenance.k8s.io` finalizer. 

#### Pod Selection

The pods for evacuation would first be selected by node (`.spec.nodeSelector`). NodeMaintenance
should eventually remove all the pods from each node. To do this in a graceful manner, the
controller will first ensure that lower priority pods are evacuated first for the same pod type.
The user can also target some pods earlier than others with a label selector.

> DaemonSet and Static pods typically run critical workloads that should be scaled down last.

Reviewer 13: ```suggestion
DaemonSet and static pods typically run critical workloads that should be scaled down last.
```

`.spec.drainPlan` among them for the intersecting nodes. Non-intersecting nodes may have a higher
priority or pod type. The next entry in the plan can be selected once all the nodes of a
NodeMaintenance have finished evacuation and all the NodeMaintenances of intersecting nodes have
finished evacuation for the current drain targets. See the [Pod Selection and DrainTargets Example](#pod-selection-and-draintargets-example)
for additional details.

A similar kind of drain plan, albeit with fewer features is offered today by the
[Graceful Node Shutdown](https://kubernetes.io/docs/concepts/architecture/nodes/#graceful-node-shutdown)
feature and by the Cluster Autoscaler's [drain-priority-config](https://github.com/kubernetes/autoscaler/pull/6139).
The downside of these configurations is that they have `shutdownGracePeriodSeconds` which sets a
limit on how long the termination of pods should take. This is not application-aware and some
applications may require more time to gracefully shut down. Allowing such hard-coded timeouts may
result in unnecessary application disruptions or data corruption.

> To support the evacuation of `DaemonSet` and `Static` pods, the daemon set controller and kubelet

Reviewer 13: ```suggestion
To support the evacuation of `DaemonSet` and `Static` pods, the DaemonSet controller and kubelet
```

should observe NodeMaintenance objects and Evacuations to coordinate the scale down of the pods on
the targeted nodes.

To ensure more streamlined experience we will not support the default kubectl [drain filters](https://github.com/kubernetes/kubernetes/blob/56cc5e77a10ba156694309d9b6159d4cd42598e1/staging/src/k8s.io/kubectl/pkg/drain/filters.go#L153-L162).
Instead, it should be possible to create the NodeMaintenance object with just a `spec.nodeSelector`.
The only thing that can be configured is which pods should be scaled down first.

NodeMaintenance alternatives to kubectl drain filters:

> - `daemonSetFilter`: Removal of these pods should be supported by the daemonset controller.

Reviewer 13: ```suggestion
- `daemonSetFilter`: Removal of these pods should be supported by the DaemonSet controller.
```

- NodeMaintenance condition or annotation can be set on the node object to advertise the current
  phase of the maintenance.

#### Supported Stage Transitions

The following transitions should be validated by the API server.

- Idle -> _Deletion_
  - Planning a maintenance in the future and canceling/deleting it without any consequence.
- (Idle) -> Cordon -> (Complete) -> _Deletion_.
  - Make a set of nodes unschedulable and then schedulable again.
  - The complete stage will always be run even without specifying it.
- (Idle) -> (Cordon) -> Drain -> (Complete) -> _Deletion_.
  - Make a set of nodes unschedulable, drain them, and then make them schedulable again.
  - Cordon and Complete stages will always be run, even without specifying them.
- (Idle) -> Complete -> _Deletion_.
  -  Make a set of nodes schedulable.

The stage transitions are invoked either manually by the cluster admin or by a higher-level
controller. For a simple drain, cluster admin can simply create the NodeMaintenance with
`stage: Drain` directly.

### DaemonSet Controller

The DaemonSet workloads should be tied to the node lifecycle because they typically run critical

> workloads where availability is paramount. Therefore, the daemonset controller should respond to

Reviewer 13: ```suggestion
workloads where availability is paramount. Therefore, the DaemonSet controller should respond to
```

the Evacuation only if there is a NodeMaintenance happening on that node and the DaemonSet is in
the `drainTargets`. For example, if we observe the following NodeMaintenance:

```yaml
apiVersion: v1alpha1
kind: NodeMaintenance
...
status:
  nodeStatuses:
    - nodeRef:
        name: six
      drainTargets:
        - podPriority: 2147483647
          podType: Default
        - podPriority: 5000
          podType: DaemonSet
  ...
```

> To register the daemonset controller as a controller evacuator, it should set the following

Reviewer 13: ```suggestion
To register the DaemonSet controller as a controller evacuator, it should set the following
```

annotation on its own pods.

```yaml

> evacuation.coordination.k8s.io/daemonset.apps.k8s.io: "10000/controller"

Reviewer: This KEP should make it clear that the implementer needs to register this annotation (or set of annotations).

Author: updated to:

```
To fulfil the Evacuation API, the daemonset controller should register itself as a controller
evacuator. To do this, it should ensure that the following annotation is present on its own pods.
```

is this  better?

```

The controller should respond to the Evacuation object when it observes its own class
(`daemonset.apps.k8s.io`) in `.status.activeEvacuatorClass`.

For the above node maintenance, the controller should not react to Evacuations of DaemonSet pods
with a priority greater than 5000. This state should not normally occur, as Evacuation requests
should be coordinated with NodeMaintenance. If it does occur, we should not encourage this flow by
updating the `.status.ActiveEvacuatorCompleted` field, although it is required to update this field
for normal workloads.

If the DaemonSet pods have a priority equal to or less than 5000, the Evacuation status should be

> updated appropriately as follows, and the targeted pod should be deleted by the daemonset

Reviewer 13: ```suggestion
updated appropriately as follows, and the targeted pod should be deleted by the DaemonSet
```

controller:

```yaml
apiVersion: v1alpha1
kind: Evacuation
metadata:
  finalizers:

>     evacuation.coordination.k8s.io/instigator_nodemaintenance.k8s.io

Reviewer: Do we typically use `_` in finalizer names? It doesn't quite fit what I'd expect.

Author: Please see https://github.com/kubernetes/enhancements/pull/4213#discussion_r1600458460 for an explanation.

FYI, we use finalizers:
- To ensure that only the instigator can cancel the evacuation and that Evacuations are not stomped over/deleted by other parties.
- To track multiple instigators.
- To allow reaction to deletions by the instigator and the evacuator. Most likely they will want to atomically complete the evacuation. 
- It is in this form for user convenience so we can automatically remove Evacuations when the pod is removed.

for additional information, please see
- https://github.com/kubernetes/enhancements/blob/2d7dbab6c95575d68395c30fa789f87d9a2c3c01/keps/sig-apps/4563-evacuation-api/README.md#evacuation-instigator-1
- https://github.com/kubernetes/enhancements/blob/2d7dbab6c95575d68395c30fa789f87d9a2c3c01/keps/sig-apps/4563-evacuation-api/README.md#evacuation-instigator-finalizer

  labels:
    app: critical-ds
  name: ae9b4bc6-e4ca-4f8e-962b-2d4459b1f684-critical-ds-5nxjs
  namespace: critical-workloads
spec:
  podRef:
    name: critical-ds-5nxjs
    uid:  ae9b4bc6-e4ca-4f8e-962b-2d4459b1f684
  progressDeadlineSeconds: 1800
  evacuators:
    - evacuatorClass: daemonset.apps.k8s.io
      priority: 10000
      role: controller
status:
  activeEvacuatorClass: daemonset.apps.k8s.io
  activeEvacuatorCompleted: false
  evacuationProgressTimestamp: "2024-04-22T21:40:32Z"
  expectedEvacuationFinishTime: "2024-04-22T21:41:32Z" # now + terminationGracePeriodSeconds:
  failedEvictionCounter: 0
  message: "critical-ds is terminating the pod due to node maintenance (OS upgrade)."
  conditions: []
```

Once the pod is terminated and removed from the node, it should not be re-scheduled on the node by

> the deamonset controller until the node maintenance is complete.

Reviewer 13: ```suggestion
the DeamonSet controller until the node maintenance is complete.
```

Reviewer: Actually DaemonSet

Reviewer 13: @sftim, thank you!

> ### Kubelet: Graceful Node Shutdown

Reviewer 13: ```suggestion
### kubelet: Graceful Node Shutdown
```

The current Graceful Node Shutdown feature has a couple of drawbacks when compared to

> NodeMaintenance:

Reviewer 9: graceful/non-graceful shutdown and NodeMaintenance procedures on a node could interfere.
As suggested [previously](https://github.com/kubernetes/enhancements/pull/4213#pullrequestreview-1971258767), NMO has achieved mutual exclusion from NHC (and node remediators) by using node leases that could hint who is doing remediation/maintenance so they both won't happen simultaneously and be agnostic.

These two procedures are important for a faulty node, thus it should be considered how they would work when a user wants them together:

- Blocking one of the procedures while one is running (/Active)
- Block and wait until one of the procedures is finished (so no interference)
- Modify the procedures so they can work together (sequentially) without interference

Author: @razo7 When I look at the NMO and NHC, the important part they solve is the node lifecycle/remediation (among other things).

I am not sure if NodeMaintenance's responsibility should be the node lifecycle. To check if the node is healthy or down. If the upgrade is in progress or not. If the shutdown is in progress or has finished. Etc.

Also, automatically triggering the shutdown after the node has drained, seems like a tricky topic.

I think it would be better to leave this responsibility to other actors (e.g. node lifecycle components) that can use the NodeMaintenance and then react to its status accordingly. One of these components is also kubelet, as described in the KEP. 

Initially, NodeMaintenance should focus on properly solving the node drain. To tell the applications how they should leave the node. In what order and safely.

We can discuss supporting additional features by NodeMaintenance later, after some adoption, if needed.

Author: Regarding the race for the node drain. NodeMaintenance should potentially solve all the important use cases.
And having multiple NodeMaintenance objects from different actors is already supported in the KEP.

Also, since this will be the most graceful solution for the node drain, we should not clash with other more aggressive drainers. (*although NodeMaintenance can be more aggressive for workloads guarded by PDBs).

Today's different types of evictions, deletions and drainers also do not synchronize who is removing what workload and when.

Reviewer: For me, NodeMaintenance would provide three things:
- a coordination point for clients of the Kubernetes API to understand more about why a node might be tainted, out of service, etc
- a building block for additional, out-of-project controllers and operators
  - for example, make an owned NodeMaintenance that belongs to your ClusterKernelPatchRequest custom resource.
  - for example, automatically create a NodeMaintenance in response to an external API marking the **antarctica-d02** zone as down, matching all affected nodes by label. Remove it once you've sailed over and checked the cluster.
- server-side tracking and control for drains, `kubectl drain` style

We shouldn't try to provide too much of the specifics, so long as the basic thing doesn't make those use cases particularly hard to implement one day in the future.

Reviewer: I can imagine a NodeMaintenance initiated as a result of a UPS reporting it has at most 3s of power left (because the UPS is overheating). I can also imagine scheduling node maintenance months in advance. There are more user stories than we have room to track.

Author: I have added these under the non goals: https://github.com/kubernetes/enhancements/blob/a15b78d3b4f8bc1b4a4e5b1508e4ad733e64c4fd/keps/sig-apps/4212-declarative-node-maintenance/README.md#non-goals to at least track this somewhere.

Also see #8 in [Motivation](https://github.com/kubernetes/enhancements/blob/a15b78d3b4f8bc1b4a4e5b1508e4ad733e64c4fd/keps/sig-apps/4212-declarative-node-maintenance/README.md#motivation) as @kwilczynski noted in https://github.com/kubernetes/enhancements/pull/4213#discussion_r1598305160. There are problems today, with the pod termination synchronization.

- It is application agnostic as it only provides a static grace period before the shutdown based on
  priority. This does not always give the application enough time to react and can lead to data
  loss or application availability loss.
- The DaemonSet pods may be running important services (critical priority) that should be available

>   even during part of the shutdown. The daemon set controller does not have the observability of

Reviewer 13: A recently filled issue also highlighted a problem between the Taint manager and Graceful Node Shutdown, which, similarly to the problem with the DaemonSet controller, can potentially interfere with the graceful termination. See:

- https://github.com/kubernetes/kubernetes/issues/124448
- https://github.com/kubernetes/kubernetes/issues/72129

Lack of information or internet (if you wish) exchange between Graceful Node Shutdown manager and rest of the platform is definitely something that seems to be a problem more often than not.

Author: +1, I haved added a mention of these under `#8`  in [Motivation](https://github.com/kubernetes/enhancements/blob/a15b78d3b4f8bc1b4a4e5b1508e4ad733e64c4fd/keps/sig-apps/4212-declarative-node-maintenance/README.md#motivation). Also mentioned in the [non-goals](https://github.com/kubernetes/enhancements/blob/a15b78d3b4f8bc1b4a4e5b1508e4ad733e64c4fd/keps/sig-apps/4212-declarative-node-maintenance/README.md#non-goals).

I am not really sure if we can stop pod termination mechanisms (e.g. taint eviction) for pods that are under NodeMaintenance/Evacuation. Seems like a backwards incompatible change. @kwilczynski @sftim thoughts?

  the kubelet shutdown procedure and cannot infer which DaemonSets should stop running. The
  controller needs to know which DaemonSets should run on each node with which priorities and
  reconcile accordingly.

To support these use cases we could introduce a new configuration option to the kubelet called
`preferNodeMaintenanceDuringGracefulShutdown`.

This would result in the following behavior:

When a shutdown is detected, the kubelet would create a NodeMaintenance object for that node.
Then it would block the shutdown indefinitely, until all the pods are terminated. The kubelet
could pass the priorities from the `shutdownGracePeriodByPodPriority` to the NodeMaintenance,
just without the `shutdownGracePeriodSeconds`. This would give applications a chance to react and
gracefully leave the node without a timeout. [Pod Selection](#pod-selection) would ensure that user
workloads are terminated first and critical pods are terminated last. 

By default, all user workloads will be asked to terminate at once. The Evacuation API ensures that
an evacuator is selected or an eviction API is called. This should result in a fast start of a pod

> termination.NodeMaintenance could then be used even by spot instances.

Reviewer 13: ```suggestion
termination. NodeMaintenance could then be used even by spot instances.
```

The NodeMaintenance object could survive kubelet restarts, and the kubelet would always know if the

> node is under shutdown (maintenance). The cluster admin would have to remove the NodeMaintenance

Reviewer 13: Would the shutdown cancellation work in the same way as it currently is under Graceful Node Shutdown? I assume, yes, it would.

Reviewer: This KEP also supports node maintenance for Windows nodes, where graceful shutdown handling is (AIUI) not implemented.

Reviewer: IMO you can mark a node as needing maintenance even if it doesn't run systemd (either Linux without systemd, or Windows which doesn't really have an init process in the POSIX sense). So this KEP can move forward whether or not your nodes gracefully handle a shutdown signal.

Author: [Cancelation by kubelet](https://kubernetes.io/docs/concepts/cluster-administration/node-shutdown/#graceful-node-shutdown)
>  Note:
There are cases when Node termination was cancelled by the system (or perhaps manually by an administrator). In either of those situations the Node will return to the Ready state. However, Pods which already started the process of termination will not be restored by kubelet and will need to be re-scheduled.

Is the node termination cancellation reliable? Can we use it for deletion of the NodeMaintenance?

---

Also the behavior is application specific, when the the NodeMaintenance is cancelled/deleted (please see the [Evacuation KEP](https://github.com/kubernetes/enhancements/pull/4565)). The application/evacuator can  terminate the pod or it can also recover.

object after the node restart to indicate that the node is healthy and can run pods again. Admins
are expected to deal with the lifecycle of planned NodeMaintenances, so reacting to the unplanned
one should not be a big issue.

If there is no connection to the apiserver (apiserver down, network issues, etc.) and the
NodeMaintenance object cannot be created, we would fall back to the original behavior of Graceful

> Node Shutdown feature. If the connection is restored, we would stop the Graceful Node Shutdown and

Reviewer 13: It might be too late to fallback to something else. We should attempt to add NodeMaintenance object, should it be missing, but it might be too late to stop Graceful Node Shutdown, which might be too far gone, so to speak. This would be a one-way? hard? fallback.

Reviewer 13: Alternatively, perhaps then doing nothing or refusing to carry-out a shutdown would be a safer option?

That said, depends on how the shutdown was triggered: if it was due to external factors (which would be the default modus operandi for which Graceful Node Shutdown has been designed), then it would be potentially unsafe to interrupt the ongoing shutdown, or there might be not enough time, etc.

Author: It would be too late for the pods that are already terminating. But it might be useful for the running non-terminated pods, right?

It seems to me that by default it would be good to carry out the Graceful Node Shutdown in the disconnected scenario. At least to have a softer landing in bad situations. I wonder if it would make sense to make this behavior configurable?

Reviewer: AIUI, DNM is about marking the node as needing maintenance and barely overlaps with node shutdown handling.

The only overlap is that both things trigger a node drain; but other things could also trigger that, such as node problem detector deciding the node is irreparably faulty.

proceed with the NodeMaintenance.

The NodeMaintenance would ensure that all pods are removed. This also includes the DaemonSet and

> Static pods.

Reviewer 13: ```suggestion
Static Pods.
```

> ### Kubelet: Static Pods

Reviewer 13: ```suggestion
### kubelet: Static Pods
```

> Currently, there is no standard solution for terminating static pods. We can advertise what state

Reviewer 13: ```suggestion
Currently, there is no standard solution for terminating Static Pods. We can advertise what state
```

Reviewer: “static pods” is fine, IMO.

Reviewer 13: I went with the capitalisation, given that it's the name of a feature. If you think it's fine, I will then delete the suggestions to remove some of the noise (for which I apologise in advance).

Reviewer: We [use UpperCamelCase for API kinds](https://kubernetes.io/docs/contribute/style/style-guide/#use-upper-camel-case-for-api-objects), but then write some kinds in lower case when we're talking about the thing the object represents more than the API object.

> When you are generally discussing an API object, use [sentence-style capitalization](https://docs.microsoft.com/en-us/style-guide/text-formatting/using-type/use-sentence-style-capitalization).

Because static pods don't appear in the API (their mirror pods do), and because “static Pod” jars more than “static pod”, I'd use all-lowercase.

Reviewer 13: @sftim, nice! Good to know for the future reference. Appreciated. :+1:

> each node should be in, declaratively with NodeMaintenance. This can include static pods as well.

Reviewer 13: ```suggestion
each node should be in, declaratively with NodeMaintenance. This can include Static Pods as well.
```

> Since static pods usually run the most critical workloads, they should be terminated last according

Reviewer 13: ```suggestion
Since Static Pods usually run the most critical workloads, they should be terminated last, according
```

to [Pod Selection](#pod-selection).

> Similar to [DaemonSets](#daemonset-controller), Static pods should be tied to the node lifecycle

Reviewer 13: ```suggestion
Similar to [DaemonSets](#daemonset-controller), Static Pods should be tied to the node lifecycle
```

because they typically run critical workloads where availability is paramount. Therefore, the
kubelet should respond to the Evacuation only if there is a NodeMaintenance happening on that node

> and the Static pod is in the `drainTargets`. For example, if we observe the following

Reviewer 13: ```suggestion
and the Static Pod is in the `drainTargets`. For example, if we observe the following
```

NodeMaintenance:

```yaml
apiVersion: v1alpha1
kind: NodeMaintenance
...
status:
  nodeStatuses:
    - nodeRef:
        name: six
      drainTargets:
        - podPriority: 2147483647
          podType: Default
        - podPriority: 2147483647
          podType: DaemonSet
        - podPriority: 7000
          podType: Static
  ...
```

To register the kubelet as a controller evacuator, it should set the following annotation on the
mirror pods.

```yaml
evacuation.coordination.k8s.io/kubelet.k8s.io: "10000/controller"
```

The kubelet should respond to the Evacuation object when it observes its own class
(`kubelet.k8s.io`) in `.status.activeEvacuatorClass`.

> 

Reviewer 13: ```suggestion
```

> For the above node maintenance, the kubelet should not react to Evacuations of Static pods

Reviewer 13: ```suggestion
For the above node maintenance, the kubelet should not react to Evacuations of Static Pods
```

with a priority greater than 7000. This state should not normally occur, as Evacuation requests
should be coordinated with NodeMaintenance. If it does occur, we should not encourage this flow by
updating the `.status.ActiveEvacuatorCompleted` field, although it is required to update this field
for normal workloads.

> If the static pods have a priority equal to or less than 7000, the Evacuation status should be

Reviewer 13: ```suggestion
If the Static Pods have a priority equal to or less than 7000, the Evacuation status should be
```

- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

## Alternatives

### Out-of-tree Implementation

We could implement the NodeMaintenance or Evacuation API out-of-tree first as a CRD.

One of the problems is that it would be difficult to get real word adoption and thus important
feedback on this feature. This is mainly due to the requirement that the feature be implemented
and integrated with multiple components to observe the benefits. And those components are both
cluster admin and application developer facing.

There is a [Node Maintenance Operator](https://github.com/medik8s/node-maintenance-operator) project
that provides a similar API and has some adoption. But it is not at a level where applications could
depend on this API to be present in the cluster. So it doesn't make that much sense to implement
the app migration logic as it cannot be applied everywhere. As shown in the motivation part
of this KEP, there is a big appetite for having a unified and stable API that could be used by
everyone to implement the new capabilities.

> ### Use a Node Object Instead of Introducing a New NodeMaintenance API

Reviewer 13: ```suggestion
### Use a Node Object Instead of Introducing a New NodeMaintenance API

```

`.spec.tolerations[].tolerationSeconds`, which would give a deadline for the pods to be removed by
the NoExecuteTaintManager.

This approach has the following disadvantages:
- Taints and tolerations do not support PDBs, which is the main mechanism for preventing voluntary
  disruptions. People who want to avoid the disruptions caused by the maintenance taint would have
  to specify the toleration in the pod definition and ensure it is present at all times. This would
  also have an impact on the controllers, who would have to pollute the pod definitions with these
  tolerations, even though the users did not specify them in their pod template. The controllers
  could override users' tolerations, which the users might not be happy about. It is also hard to
  make such behaviors consistent across all the controllers.
- Taints are used as a mechanism for involuntary disruption; to get pods out of the node for some
  reason (e.g. node is not ready). Modifying the taint mechanism to be less harmful
  (e.g. by adding a PDB support) is not possible due to the original requirements.
- It is not possible to incrementally scale down according to pod priorities, labels, etc.

> ### Names considered for the new API

Reviewer 13: ```suggestion
### Names considered for the new API

```

These names are considered as an alternative to NodeMaintenance:

- NodeIsolation
- NodeDetachment
- NodeClearance
- NodeQuarantine
- NodeDisengagement
- NodeVacation

> 

Reviewer 13: ```suggestion
```

    - Testing: Are there any tests for failure mode? If not, describe why.
-->

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

## Alternatives

### Out-of-tree Implementation

We could implement the NodeMaintenance or Evacuation API out-of-tree first as a CRD.

One of the problems is that it would be difficult to get real word adoption and thus important
feedback on this feature. This is mainly due to the requirement that the feature be implemented
and integrated with multiple components to observe the benefits. And those components are both

> cluster admin and application developer facing.

Reviewer 14: Using the Core APIs to get adoption is not a good argument, if we are assuming there is appetite official CRDs will be equal desired for the users ... adding experimental features intersecting with the Pod and Node lifecycle than later does not get adoption is a big penalty for the project

Author: I have changed the section to better reflect the current state of the KEP. Mainly the technical reasons why we prefer the core API. I have kept the adoption part there, but slimmed it down. The adoption was there from the time we had the Evacuation API in the KEP, but since we moved it to a separate KEP, it is not as important anymore.

   then attempt to perform a [Graceful Node Shutdown](https://kubernetes.io/docs/concepts/architecture/nodes/#graceful-node-shutdown):
   - delay the shutdown pending graceful termination of remaining pods
   - terminate remaining pods in reverse priority order (see [pod-priority-graceful-node-shutdown](https://kubernetes.io/docs/concepts/architecture/nodes/#pod-priority-graceful-node-shutdown))

The main problem is that the current approach tries to solve this in an application agnostic way
and will simply attempt to remove all the pods currently running on the node. Since this approach
cannot be applied generically to all pods, the Kubernetes project has defined special
[drain filters](https://github.com/kubernetes/kubernetes/blob/56cc5e77a10ba156694309d9b6159d4cd42598e1/staging/src/k8s.io/kubectl/pkg/drain/filters.go#L153-L162)
that either skip groups of pods or an admin has to consent to override those groups to be either
skipped or deleted. This means that without knowledge of all the underlying applications on the
cluster, the admin has to make a potentially harmful decision.

From an application owner or developer perspective, the only standard tool they have is

> a PodDisruptionBudget. This is sufficient in a basic scenario with a simple multi-replica

Reviewer 5: Do we have a list of all PDB issues somewhere?

Author: I think the main ones are in the KEP, but I can add the rest of them for the context .

application. The edge case applications, where this does not work are very important to
the cluster admin, as they can block the node drain. And, in turn, very important to the
application owner, as the admin can then override the pod disruption budget and disrupt their
sensitive application anyway.

List of cases where the current solution is not optimal:

> 1. Without extra manual effort, an application running with a single replica has to settle for

Author: Some applications have a general ability (e.g. deployments) to scale up first. So they do not have to experience the downtime if configured correctly. This can be generalized to any number of replicas. There are more details in the linked issues.

   experiencing application downtime during the node drain. They cannot use PDBs with
   `minAvailable: 1` or `maxUnavailable: 0`, or they will block node maintenance. Not every user
   needs high availability either, due to a preference for a simpler deployment model, lack of
   application support for HA, or to minimize compute costs. Also, any automated solution needs
   to edit the PDB to account for the additional pod that needs to be spun to move the workload
   from one node to another. This has been discussed in the issue [kubernetes/kubernetes#66811](https://github.com/kubernetes/kubernetes/issues/66811)
   and in the issue [kubernetes/kubernetes#114877](https://github.com/kubernetes/kubernetes/issues/114877).
2. Similar to the first point, it is difficult to use PDBs for applications that can have a variable

>    number of pods; for example applications with a configured horizontal pod autoscaler (HPA). These

Reviewer 5: The two possible alternative would be:
1. extend HPA to update associated PDB to maintain the required minimum? 
2. every application should have a reasonable minimum which would allow the application to survive any kind of disruption (planned or unplanned).

Author: > extend HPA to update associated PDB to maintain the required minimum?

- This is not possible when n=1, which is the core of the https://github.com/kubernetes/kubernetes/issues/93476 issue.
- Although it is possible for n>1, I would be afraid of conflicts when introducing such feature. PDB was not introduced as a controlled/changing object like for example ReplicaSets. It is set by operators/controllers, manually or as part of various devops flows. 
- There is no backlink from pod/HPA to a PDB. Even if we introduce such a backlink, there are still problems. What do we do when the PDB does not exist and there is one specified by a HPA. Should it be the responsibility of the HPA to become an operator of a PDB (or become other type of a PDB)?  We could run again into the conflicts with existing flows as seen in the previous point.
- There is no mapping from a PDB to an application, there is only a selector. Although not recommended, there could be two applications under one PDB. One controlled by HPA and the other one not. Or any combination of these.

> every application should have a reasonable minimum which would allow the application to survive any kind of disruption (planned or unplanned).

This is not always possible/desired, as seen in https://github.com/kubernetes/kubernetes/issues/66811, https://github.com/kubernetes/kubernetes/issues/114877 and https://github.com/kubernetes/kubernetes/issues/93476

   applications cannot be disrupted during a low load when they have only pod. However, it is

>    possible to disrupt the pods during a high load without experiencing application downtime. If

Reviewer 5: I doubt any admin is interested in node maintenance during high load, unless in a critical situation, in which case I'm not sure any of the considerations from this document might apply, because you're focusing on bringing the cluster back to living.

Author: High load (of the application) is just another way of saying that the HPA has decided to scale beyond 1 pod. That does not have to correlate to the high load of the cluster.

   This has been discussed in issues [kubernetes/kubernetes#122674](https://github.com/kubernetes/kubernetes/issues/122674),
   [kubernetes/kubernetes#120613](https://github.com/kubernetes/kubernetes/issues/120613) and [kubernetes/kubernetes#122674](https://github.com/kubernetes/kubernetes/issues/112443).
7. There is no way to gracefully terminate static pods during a node shutdown
   [kubernetes/kubernetes#122674](https://github.com/kubernetes/kubernetes/issues/122674), and the
   lifecycle/termination is not clearly defined for static pods [kubernetes/kubernetes#16627](https://github.com/kubernetes/kubernetes/issues/16627).
8. Different pod termination mechanisms are not synchronized with each other. So for example, the
   taint manager may prematurely terminate pods that are currently under Node Graceful Shutdown.
   This can also happen with other mechanism (e.g., different types of evictions). This has been
   discussed in the issue [kubernetes/kubernetes#124448](https://github.com/kubernetes/kubernetes/issues/124448)
   and in the issue [kubernetes/kubernetes#72129](https://github.com/kubernetes/kubernetes/issues/72129).

> 9. There is not enough metadata about why the node drain was requested. This has been discussed in

Reviewer 5: A lot of the above points nicely collected issues with Graceful Node Shutdown, which is being worked on as part of https://github.com/kubernetes/enhancements/issues/2000. Have you considered pointing the authors to this list?

Author: There has been a cooperation/discussion with sig-node. I have discussed this with @kwilczynski, who is interested in graduating the Graceful Node Shutdown feature. 

Many of the problems described are impossible or impractical to be solved by the kubelet alone. This is one of the main goals of this KEP to help resolve this situation.

cc @SergeyKanzhelev @dchen1107

  (the check can be modified with the `skip-nodes-with-system-pods` option).
- Pods with `cluster-autoscaler.kubernetes.io/safe-to-evict: "false"` annotation are blocking.
- Pods with local storage are blocking unless they have a
  `cluster-autoscaler.kubernetes.io/safe-to-evict-local-volumes` annotation
  (the check can be modified with the `skip-nodes-with-local-storage` option, e.g., this check
  is skipped on [AKS](https://learn.microsoft.com/en-us/azure/aks/cluster-autoscaler?tabs=azure-cli#cluster-autoscaler-profile-settings)).
- Pods with PDBs that do not have `disruptionsAllowed` are blocking.

This can be enhanced with other rules and overrides.

It uses this logic to first check if all pods can be removed from a node. If not, it will report
those nodes. Then it will group all pods by a priority and evict them gradually from the lowest
to highest priority. This may include DaemonSet pods.

### kubelet

Graceful Node Shutdown is a part of the current solution for node maintenance. Unfortunately, it is
not possible to rely solely on this feature as a go-to solution for graceful node and workload
termination.

> - The Graceful Node Shutdown feature is not application aware and may prematurely disrupt workloads 

Reviewer 5: None of the solutions will be ever application aware, that is one of the reasons we have operators per a specific application, which know how to deal with all the intricacies of managing it.

Author: Yes, I am only stating the fact. Not implying that it should be. The point of this KEP is to freely share the intentions of draining a node and each pod so that each actor (controller, application, user, tools, etc.) can react in a proper way.

  and lead to data loss.
- The kubelet controls the shutdown process using Dbus and systemd, and can delay (but not entirely

>   block) it using the systemd inhibitor. However, if Dbus or the kubelet is restarted during the

Reviewer 5: That sounds like a bug in the kubelet, no?

Author: AFAIK this is an inherent problem of the current design and not something that can be solved entirely by kubelet. To solve this they need to store the intent, which would be done through NodeMaintenance in this proposal.

Reviewer 14: if kubelet is restarted how the intent will be stored? how it will reconcile after start?

Author: It would first have to reconnect to the apiserver, to obtain the NodeMaintenance or [Node](https://github.com/kubernetes/enhancements/pull/4213#discussion_r1631444470) object.

  node shutdown, the shutdown might not be registered again, and pods might be terminated

>   ungracefully. Also, new workloads can get scheduled on the node while the node is shutting down.

Reviewer 5: This seems more like a mismanaged node maintenance, that no solution can fix for you, other than having a strict runbook you follow when performing one.

Reviewer: Adding an API to mark that the node is covered by a maintenance period should help here; it'd be a declarative way for not just tainting nodes but also allowing workload controllers to surge before the taint actually triggers eviction. The API I'd imagined provides more context than a taint on its own can.

Either the scheduler pays attention to node maintenance or learns to take account of a derived taint, but either way there should be a way to either prefer or enforce that new Pods don't get bound onto a node due to shut down imminently.

Reviewer 14: I don't understand why the indirection on a third API object it is better, can not we add this state to the node object itself? we already have taints and conditions that indicate the node is not ready or it can not receive pods

Reviewer: Drawbacks of adding to Node:
- one of the ideas was to define a _selector_ matching multiple nodes under maintenance. If you put it into the `.spec` of the Node, then something has to set / manage that spec
- you can have two or more concurrent reasons to mark a node as being liable to disruption
  1. example: the node autoscaler for that cluster no longer wants the node to exist
  2. also, that zone is being emptied in response to a provider-level maintenance notification
  3. the node problem detector has also triggered a fault report on this node
  
  using Server-Side Apply and multiple managers on fields is feasible, but puts the onus on each client to play well with others.
- we can revise the implementation at alpha and rely on clients to only integrate with implementations they know they support, using a match on the API version(s) served for NodeMaintenance. If we add the API to Node, clients don't have that option. The Node API will be available at v1 indefinitely and the meaning of a particular field is stable the moment we add it, pretty much.
- with a taint, we can't record a reason. If we add an official equivalent of NodeMaintenance from https://github.com/medik8s/node-maintenance-operator, we can.
  - I'm inclined to offer optional start and end times (UTC) for maintenance, because that matches real-world needs.

Reviewer 14: > one of the ideas was to define a selector matching multiple nodes under maintenance. If you put it into the .spec of the Node, then something has to set / manage that spec

you are defining a state on the node, spec seems the right place

> you can have two or more concurrent reasons to mark a node as being liable to disruption

exactly there is only one Node and one state, so is the object that provides the synchronization, I may miss the use case of playing well together, the lifecycle of the node should respect a state machine

> we can revise the implementation

Node should only have the necessary primitives ... and fix the bugs and limitations that exist today, that need some work  as @soltysh mentioned in other comments .Building compensation layers is easier in the short term but ends adding more technical debt,

 if the lifecycle of the nodes and the primitives implemented on the Node exist, we don't need to absorb the complexity of an orchestrator in tree with all the actuators and evacuators and .... 

> with a taint, we can't record a reason. If we add an official equivalent of NodeMaintenance from https://github.com/medik8s/node-maintenance-operator, we can.

> I'm inclined to offer optional start and end times (UTC) for maintenance, because that matches real-world needs.

This start to move to the area of orchestration and day 2 operations, does not feel something for core, and more for ClusterAPI or another projects on that area

Author: It is hard to achieve synchronization of a state through a single object, if you have multiple actors requesting the maintenance and then cancelling it at different times. We are talking about admins sending patch requests, not a handful of well-behaved components. 

We have the same problem today with cordon/uncordon (`.spec.Unschedulable`) today. See https://github.com/kubernetes/enhancements/pull/4213#discussion_r1537798784. This KEP also has to handle this field in a special manner. If it becomes a problem, we can detect NM in the scheduler.

We can move some part of the state/status to the node as mentioned in https://github.com/kubernetes/enhancements/pull/4213#discussion_r1631539092. But it is still useful to have the NodeMaintenance API to decouple the ownership of the maintenance and also to decouple the functional aspects.

The evacuation API is also designed with the concurrency in mind. The API is not only useful for the NodeMaintenance, but has much broader use cases (comparable/larger than eviction). It provides a way for any actor to observe the eviction/evacuation and thus it is possible to gracefully remove any type of pod (normal pod, daemonset, static, pod under a PDB, pod not under a PDB, pod of a workload that does not support PDBs, pod in any state). Applications would have full control over the lifecycle of their own pods. This is not possible today, and people are forced to write a lot of workarounds to get closer to the desired behavior.

Later, it should be possible to extend the Evacuation API if needed, to gracefully remove any kind of a resource.

Reviewer: > you are defining a state on the node, spec seems the right place

I'm expecting to define a desired state across a _set_ of nodes (potentially 4999 of them: our soft limit on node scaling, minus one). If I can define that in one API call, it's easier than making 4999 calls and the client gets to return a single success failure.

  Cluster admins should, therefore, plan the maintenance in advance and ensure that pods are
  gracefully removed before attempting to shut down or restart the machine.
- The kubelet has no way of reliably detecting ongoing maintenance if the node is restarted in the
  meantime.
- Graceful termination of static pods during a shutdown is not possible today. It is also not
  currently possible to prevent them from starting back up immediately after the machine has been
  restarted and the kubelet has started again, if the node is still under maintenance. 

### Motivation Summary

To sum up. Some applications solve the disruption problem by introducing validating admission

> webhooks. This has some drawbacks. The webhooks are not easily discoverable by cluster admins. And

Reviewer 14: Is this not solved with the new addition of CEL?

Author: CEL cannot do everything. For example, it cannot check additional resources, call other APIs. Also the eviction API was not designed to support this, so it is still an unexpected behavior for the clients.

Reviewer 14: I see, I fully agree webhooks are something we should move away

> they can block evictions for other applications if they are misconfigured or misbehave. The

Reviewer 5: Misconfigured webhooks are generally a problem for any kind of cluster operations, and as such I don't think this makes a good argument against it. There's an explicit cautious warning in the official documentation of them https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#experimenting-with-admission-webhooks. Also, "With great power comes great responsibility" :wink:

Author: Some of the solutions today ship with these webhooks, so they are not misconfigured per se. They are part of the cluster operation. As I understand it, this was done to implement safety features that are missing in kubernetes today.

The downside is mostly the observability and breaking the eviction contract (https://github.com/kubernetes/enhancements/pull/4213#discussion_r1626319288). Also, if things go wrong these webhooks will block the node drain, which is a headache for cluster administrators.

kubelet's Graceful Node Shutdown feature is a best-effort solution for unplanned shutdowns, but
it is not sufficient to ensure application and data safety.

### Goals
- Introduce NodeMaintenance API.
- Introduce a node maintenance controller that creates evacuations.
- Deprecate kubectl drain in favor of NodeMaintenance.
- Implement NodeMaintenanceAwareKubelet feature to implement a lifecycle for static pods during a
  maintenance.
- Implement NodeMaintenanceAwareDaemonSet feature to prevent the scheduling of DaemonSet pods on
  nodes during a maintenance.

### Non-Goals
- Introduce a node maintenance period, nodeDrainTimeout (similar to [cluster-api](https://cluster-api.sigs.k8s.io/developer/architecture/controllers/control-plane)
  nodeDrainTimeout) or a TTL optional field as an upper bound on the duration of node maintenance.
  Then the node maintenance would be garbage collected and the node made schedulable again.
- Solve the node lifecycle management or automatic shutdown after the node drain is completed.
  Implementation of this is better suited for other cluster components and actors who can use the
  node maintenance as a building block to achieve their desired goals.

> - Synchronize all pod termination mechanisms (see #8 in the [Motivation](#motivation) section), so that they do

Reviewer 14: This require to synchronize the Endpoints/Services too , after a quick skim to the KEP I wonder if this approach will not hit the same problem @smarterclayton is describing here https://github.com/kubernetes/kubernetes/issues/116965#issuecomment-2113227772, the Pods should report the status to avoid traffic to be sent to them if they are not ready

Reviewer: As this is a non goal I think the KEP can comfortably find an easy way to not deliver it.

Author: 1. This KEP does not propose to solve the non-goals nor does it suggest how or whether to implement them. It is a simply a list of things that could be done.
2. Even the goals section expects additional KEPs on the node and apps side to fully examine the problem. But for the sake of simplicity, I keep the NodeMaintenanceAwareDaemonSet and NodeMaintenanceAwareKubelet features here for now.

Regarding what @smarterclayton wrote, I think we have a lot of room to solve that. By utilizing both the NodeMaintenance and Evacuation API and deciding which parts should be mandatory and which optional.

  not terminate pods under NodeMaintenance/Evacuation.

## Proposal

Most of these issues stem from the lack of a standardized way of detecting a start of the node
drain. This KEP proposes the introduction of a NodeMaintenance object that would signal an intent
to gracefully remove pods from given nodes. The intent will be implemented by the newly proposed
[Evacuation API KEP](https://github.com/kubernetes/enhancements/issues/4563), which ensures
graceful pod removal or migration, an ability to measure the progress and a fallback to eviction if
progress is lost. The NodeMaintenance implementation should also utilize existing node's
`.spec.unschedulable` field, which prevents new pods from being scheduled on such a node.

> We will deprecate the `kubectl drain` as the main mechanism for draining nodes and drive the whole

Reviewer 5: With my sig-cli hat, I will call out this wasn't discussed with that group, so before proceeding with that kind of steps you'll need their approval.

Author: This is mostly symbolic, as the NodeMaintenance API will essentially replace all of its functionality. The command itself does not have to be removed. A warning that there is a better way, might be enough.

Nevertheless, I plan to present this to sig-cli before committing to it.

Reviewer: I'd imagined that we'd keep `kubectl drain` and that by default this subcommand would make a NodeMaintenance if the API is available, otherwise fall back to client-side draining.

process via a declarative API. This API can be used either manually or programmatically by other
drain implementations (e.g. cluster autoscalers). 

To support workload migration, a new controller should be introduced to observe the NodeMaintenance
objects and then select pods for evacuation. The pods should be selected by node (`nodeSelector`)
and the pods should be gradually evacuated according to the workload they are running.
Controllers can then implement the migration/termination either by reacting to the Evacuation API
or by reacting to the NodeMaintenance API if they need more details.

> ### User Stories (Optional)

Reviewer: (nit)
```suggestion
### User Stories
```

It is optional, but we clearly have some.

downscale.

#### Story 3

Cluster or node autoscalers that take on the role of `kubectl drain` want to signal the intent to
drain a node using the same API and provide a similar experience to the CLI counterpart.

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

- This KEP depends on [Evacuation API KEP](https://github.com/kubernetes/enhancements/issues/4563).

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

A misconfigured .spec.nodeSelector could select all the nodes (or just all master nodes) in the
cluster. This can cause the cluster to get into a degraded and unrecoverable state.

> An admission plugin ([NodeMaintenance Admission](#nodemaintenance-admission)) is introduced to

Reviewer 14: But this will be unrecoverable, right? if the master nodes are not available it means will you not be able to connect to the apiserver

Reviewer: The admission restriction wouldn't prevent you marking all the nodes as under maintenance. It'd just warn you that this was a bad idea.
(I'd prefer to have a ValidatingAdmissionPolicy rather than code, but we don't have lifecycle management for this kind of thing. In other words, kubeadm can't deploy recommended ValidatingAdmissionPolicies for us, so we shouldn't assume they are in place.)

- You can mark all the nodes as being covered by maintenance (and this may well be true!)
- If you as a cluster adminstrator then go and pull out the power cords or whatever, you can try to recover by plugging them back in.
- Static Pods don't stop during a node drain, right? So even if you do mark every node as under maintenance, you may still have a control plane. Perhaps, no working cluster network and no DNS, but you still have a control plane.

Author: Yes, it can still be a valid strategy. Assuming your control plane is deployed correctly - basically that it shuts down last or runs on the side (invisible to kube).

Reviewer 15: 👀 trying to better understand impacts of static Pods drain in the kubeadm scenario

Author: As I mentioned in https://github.com/kubernetes/enhancements/pull/4213#discussion_r1781840501, you can control which component goes down first and last and use it for gradual upgrades in clusters. There would have to be a specific configuration for a kubeadm cluster to let it downscale nicely. Depending on the cluster setup (HA, load balancing).

Reviewer 16: The problem is that you can't really use the loadbalancer for the communication between kubelet and kube-apiserver.

The Kubernetes skew policy states: "kubelet must not be newer than kube-apiserver.". This means that if you configure kubelet to use a loadbalancer, during upgrades you would violate the Kubernetes skew policy (example, it can happen that a 1.31 kubelet communicates with a 1.30 kube-apiserver through the loadbalancer). That's why kubeadm started configuring kubelet to use the local kube-apiserver instead.

Reviewer 17: Maybe we should consider solving the static pods in a different proposal or a follow-up? It might be worthwhile to take an incremental approach to a problem that spans across many SIGs and has a potential of wide impact on the ecosystem and current uses

Reviewer: I like the idea of a PR against the same KEP ID that adds in treatment of static Pods.

`kubectl drain` today doesn't evict static Pods, and replacements would schedule&dagger; to the same node anyway.

Reviewer: &dagger; _no actual scheduling_

   `Drained` condition is set to `"True"` on the node maintenance object.
5. Maintenance Complete: make the nodes schedulable again once the node maintenance is done.
   Configured with `.spec.stage=Complete`.

```golang

// +enum
type NodeMaintenanceStage string

const (
// Idle does not interact with the cluster.
Idle NodeMaintenanceStage = "Idle"
// Cordon cordons all selected nodes by making them unschedulable.
Cordon NodeMaintenanceStage = "Cordon"
// Drain:
// 1. Cordons all selected nodes by making them unschedulable.
// 2. Gives an intent to drain all selected nodes by creating Evacuation objects for the
//    node's pods.
Drain NodeMaintenanceStage = "Drain"
// Complete:
// 1. Removes all Evacuation objects requested by this NodeMaintenance.
// 2. Uncordons all selected nodes by making them schedulable again, unless there is not another
//    maintenance in progress.
Complete NodeMaintenanceStage = "Complete"
)

type NodeMaintenance struct {
    ...
    Spec NodeMaintenanceSpec
    Status NodeMaintenanceStatus
}

type NodeMaintenanceSpec struct {
    // NodeSelector selects nodes for this node maintenance.
    // +required
    NodeSelector *v1.NodeSelector

>     // The order of the stages is Idle -> Cordon -> Drain -> Complete.

Reviewer 18: Be really careful with this language.  The way this is specced you have a state machine, which means any futre evolution will be very difficult.

Is it allowable for a cordon'ed node to become idle?  This state machine doesn't allow it.  I think you'd do better to describe this as a set of mutually exclusive intentions.  Modes.

Author: It is hard to offer this as modes, as it should be possible to manually switch between the states/stages. And also to have a strict transitions between them (please see [Supported Stage Transitions](https://github.com/atiratree/kube-enhancements/blob/improve-node-maintenance/keps/sig-apps/4212-declarative-node-maintenance/README.md#supported-stage-transitions)) as not all of them are valid. It is up to the admin or a controller when these transitions should occur. 

We also want to have a lifecycle for the NodeMaintenance. Once the NM completes we can clean up and recover the node. If an additional NM is required, a new object should be created.

I would be happy to hear ideas on how to improve this. 

If we decide for the current solution:
- It may be possible to add new stages into the API.
- The extensibility can be offered through a set of modifiers for each stage, as we do with the `DrainPlan`.

    - Mitigations: What can be done to stop the bleeding, especially for already
      running user workloads?
    - Diagnostics: What are the useful log messages and their required logging
      levels that could help debug the issue?
      Not required until feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.
-->

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

## Alternatives

### Out-of-tree Implementation

We could implement the NodeMaintenance or Evacuation API out-of-tree first as a CRD.

> The KEP aims to solve graceful termination of any pod in the cluster. This is not possible with a

Reviewer 5: I'm seeing this not as an alternative, but rather as a prerequisite for what this document proposes. You're building a solution on top of existing k8s components, so you first need to address the problems in the underlying pieces. I've mentioned that above when asking whether your reported the problems with Node Graceful Shutdown to sig node.

Author: This KEP would help resolve the issues around Graceful Node Shutdown: https://github.com/kubernetes/enhancements/pull/4213#discussion_r1631032740

- We would like to solve the lifecycle of static pods during a node maintenance. This means that
  static pods should be terminated during the drain according to `drainPlan`, and they should stay
  terminated after the kubelet restart if the node is still under maintenance. This requires
  integration with kubelet. See [kubelet: Static Pods](#kubelet-static-pods) for more details.
- We would like to improve the Graceful Node Shutdown feature. Terminating pods via NodeMaintenance
  will improve application safety and availability. It will also improve the reliability of the
  Graceful Node Shutdown feature. However, this also requires the kubelet to interact with a
  NodeMaintenance. See [kubelet](#kubelet) and
  [kubelet: Graceful Node Shutdown](#kubelet-graceful-node-shutdown) for more details.
- We would like to also solve the lifecycle of DaemonSet pods during the NodeMaintenance. Usually
  these pods run important or critical services. These should be terminated at the right time
  during the node drain. To solve this, integration with NodeMaintenance is required. See
  [DaemonSet Controller](#daemonset-controller) for more details.

> Also, one of the disadvantages of using a CRD is that it would be more difficult to get real-word

Reviewer 5: One of the reasons for introducing CRDs was to allow authors to iterate faster and provide feedback which can:
1. improve the core k8s;
2. allow to make an informed decision whether a particular problems is worth integrating in core or not. 

So far, the several different projects you've mentioned in this document were able to achieve the functionality you're proposing here outside of core.

Author: I see, I need to emphasize this better in the goals. The kubelet integration is an important part of this KEP. 

The other projects are mostly replicating the kubectl drain logic in a controller with some extra. If this were the only part of the KEP, then I would agree that a CRD would be preferable.

Reviewer: (I think this strengthens the case for prototyping with a CRD and then switching in-tree when we add the kubelet integration).

Reviewer 5: Correct, the whole point of out-of-tree experimentation is to shape up the API and identify what has to be done in the core, that is not possible today, when using an external solution.

have to set a toleration on their pods for this maintenance taint. Such applications could also set
`.spec.tolerations[].tolerationSeconds`, which would give a deadline for the pods to be removed by
the NoExecuteTaintManager.

This approach has the following disadvantages:
- Taints and tolerations do not support PDBs, which is the main mechanism for preventing voluntary
  disruptions. People who want to avoid the disruptions caused by the maintenance taint would have
  to specify the toleration in the pod definition and ensure it is present at all times. This would
  also have an impact on the controllers, who would have to pollute the pod definitions with these
  tolerations, even though the users did not specify them in their pod template. The controllers
  could override users' tolerations, which the users might not be happy about. It is also hard to
  make such behaviors consistent across all the controllers.
- Taints are used as a mechanism for involuntary disruption; to get pods out of the node for some
  reason (e.g. node is not ready). Modifying the taint mechanism to be less harmful
  (e.g. by adding a PDB support) is not possible due to the original requirements.

> - It is not possible to incrementally scale down according to pod priorities, labels, etc.

Reviewer 5: I don't see an alternative discussed in [SIG Apps meeting from 4/29](https://youtu.be/AKnVLLAKqWA) during which Clayton proposed extending PDBs with the missing functionality like:
- backpreassure
- supporting single-replica apps
- others? (I've asked for a list of PDBs problems above).

Author: These alternatives are discussed in the Evacuation KEP: https://github.com/atiratree/kube-enhancements/blob/improve-node-maintenance/keps/sig-apps/4212-declarative-node-maintenance/README.md#alternatives

This KEP mentions all the issues, but the implementation part is more high-level in this one.

Author: I have scheduled a discussion for the next sig-apps call to discuss these options.

-->

A misconfigured .spec.nodeSelector could select all the nodes (or just all master nodes) in the
cluster. This can cause the cluster to get into a degraded and unrecoverable state.

An admission plugin ([NodeMaintenance Admission](#nodemaintenance-admission)) is introduced to
issue a warning in this scenario.

## Design Details

### Kubectl

`kubectl drain`: as we can see in the [Motivation](#motivation) section, kubectl is heavily used
either manually or as a library by other projects. It is safer to keep the old behavior of this
command. However, we will deprecate it along with all the library functions. We can print a
deprecation warning when this command is used, and promote the NodeMaintenance. Additionally, pods
that support evacuation and have `evacuation.coordination.k8s.io/priority_${EVACUATOR_CLASS}`
annotations will block eviction requests.

`kubectl cordon` and `kubectl uncordon` commands will be enhanced with a warning that will warn
the user if a node is made un/schedulable, and it collides with an existing NodeMaintenance object.
As a consequence the node maintenance controller will reconcile the node back to the old value.

> Because of this we can make these commands noop when under the NodeMaintenance.

Reviewer 14: you mean the `NodeMaintainance` object override the `cordon`  `uncordon` commands?

Author: yes, I suppose it should also be possible to block scheduling on NodeMaintenance presence only

    // Duplicate entries are not allowed.
    // This field is immutable.
    DrainPlan []DrainPlanEntry

    // Reason for the maintenance.
    Reason string
}

const (
// Default selects all pods except DaemonSet and Static pods.	
Default PodType = "Default"
// DaemonSet selects DaemonSet pods.
DaemonSet PodType = "DaemonSet"
// Static selects static pods.
Static PodType = "Static"
)

type DrainPlanEntry struct {
    // PodSelector selects pods according to their labels.
    // This can help to select which pods of the same priority should be evacuated first.
    // +optional
    PodSelector *metav1.LabelSelector
    // PodPriority specifies a pod priority.
    // Pods with a priority less or equal to this value are selected.
    PodPriority int32
    // PodType selects pods according to the pod type:
    // - Default selects all pods except DaemonSet and Static pods.	
    // - DaemonSet selects DaemonSet pods.
    // - Static selects static pods.
    PodType PodType
}

type NodeMaintenanceStatus struct {
    // StageStatuses tracks the statuses of started stages.
    StageStatuses []StageStatus
    // List of a maintenance statuses for all nodes targeted by this maintenance.

>     NodeStatuses []NodeMaintenanceNodeStatus

Reviewer 14: @thockin I do not like the pattern of using status as a database

how do we offer consistency , is there only one actor managing this? 
If the status is wiped out by mistake are we able to rebuild the current state?

Reviewer 14: is this going to support 5000 nodes per example?

Reviewer: >  If the status is wiped out by mistake are we able to rebuild the current state?

- Kubernetes no longer expects `.status` to be rebuildable
- I think it is reconstructable anyway; start from scratch (as if the kubelet had not previously been running)

Reviewer: > is this going to support 5000 nodes per example?

Good question.
We could add fields into Node status if we want the kubelet to report something per-node (and just report it, not using `.status` to record cluster administrator intent.
We could also add a SomethingSlice API like we did with Endpoints, but that does not feel like a good fit.

Author: +1, this is problematic:

I am trying to bend the status somehow, but it is not feasible.

We cannot even limit how many nodes the selectors should select on admission, since new nodes can appear during the NodeMaintenance lifetime (especially Idle). 

We could introduce a MultiNodeMaintenance object, whose only point would be to create x number of NodeMaintenance objects, one for each node. And it would synchronize the current stage between itself and the children.

But yeah, storing this in the Node status probably sounds like the best solution.

Author: This should be solved in the current API design.

    // them are selected and set for all the NodeMaintenances for that node. Thus, the DrainTargets
    // do not have to correspond to the entries in .spec.drainPlan for a single NodeMaintenance
    // instance.
    // DrainTargets cannot backtrack and will target more pods with each update until all pods on
    // the node are targeted.
    DrainTargets  []DrainPlanEntry
    // DrainMessage may specify a state of the drain on this node and a reason why the drain
    // targets are set to a particular values.
    DrainMessage string
    // Number of pods that have not yet started evacuation.
    PodsPendingEvacuation int32
    // Number of pods that have started evacuation and have a matching Evacuation object.
    PodsEvacuating int32
}

const (
    // DrainedCondition is a condition set by the node-maintenance controller that signals
    // whether all pods pending termination have terminated on all target nodes when drain is
    // requested by the maintenance object.
    DrainedCondition = "Drained"
}
```

### NodeMaintenance Admission

`nodemaintenance` admission plugin will be introduced.

> It will validate all incoming requests for CREATE, UPDATE, and DELETE operations. All nodes

Reviewer 14: on the on the `NodeMaintainance` object, right?

Author: yes, updated

> matching the `.spec.nodeSelector` must pass an authorization check for the DELETE operation.

Reviewer 14: are the nodes selected able to delete the `NodeMaintainance` object?

Reviewer: For alpha, I wouldn't have the kubelet update `.status` of NodeMaintenance only; we can add fields into Node status if we want the kubelet to report something per-node that clients need to see.

After alpha we could revisit to decide what write to NodeMaintenance spec should be done by in-tree kubelet code.

Reviewer 14: still don't clear what that means this authorization check, is kind of a nodeauthorizer check?

> After alpha we could revisit to decide what write to NodeMaintenance spec should be done by in-tree kubelet code.

what is the rationale of this decision? seems an important thing to decide, at least and unresolved block of test should be explaining this and why is unresolved

Author: > still don't clear what that means this authorization check, is kind of a nodeauthorizer check?

@aojea This is for security. So it is not enough for a person to have permissions for NodeMaintenances to drain nodes and disrupt workloads. We were asked by sig-node to include this in the KEP.

> After alpha we could revisit to decide what write to NodeMaintenance spec should be done by in-tree kubelet code.

I do not see this as a needed feature at this time. The only field that is reasonable to mutate is `stage` or maybe a `reason`. IMO it would be better to explore this in the kubelet KEPs.

> Also, if the `.spec.nodeSelector` matches all cluster nodes, a warning will be produced indicating

Reviewer 14: warning is just warning or does it block any operations?

Author: non blocking, updated

that the cluster may get into a degraded and unrecoverable state.

### NodeMaintenance Controller

Node maintenance controller will be introduced and added to `kube-controller-manager`. It will
observe NodeMaintenance objects and have the following main features:

#### Idle

> The controller should not touch the pods or nodes in any way in the `Idle` stage.

Author: yes, updated

#### Finalizers and Deletion of the NodeMaintenance

When a stage is not `Idle`, `nodemaintenance.k8s.io/maintenance-completion` finalizer is placed on
the NodeMaintenance object to ensure uncordon and removal of Evacuations upon deletion.

When a deletion of the NodeMaintenance object is detected, its `.spec.stage` is set to `Complete`.
The finalizer is not removed until the `Complete` stage has been completed.

#### Cordon

When a `Cordon` or `Drain` stage is detected on the NodeMaintenance object, the controller
will set (and reconcile) `.spec.Unschedulable` to `true` on all nodes that satisfy
`.spec.nodeSelector`. It should alert via events if too many occur appear and a race to change

> this field is detected.

Reviewer: Not sure about this but one way to handle cordon / uncordon is to label the NodeMaintenances that result from a cordon. `kubectl uncordon` could then remove any NodeMaintenance labelled as a cordon and matching the one specific node.

We either reject or warn on setting that label on a NodeMaintenance that selects or could select multiple nodes.

> this field is detected.

Author: This is supported, please see https://github.com/atiratree/kube-enhancements/blob/improve-node-maintenance/keps/sig-apps/4212-declarative-node-maintenance/README.md#pod-selection-and-draintargets-example

> this field is detected.

Author: If there is at least one maintenance with Cordon or Drain stage, the node remains unschedulable.

  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

This KEP proposes adding a declarative API to manage node maintenance. This API can be used to
implement additional capabilities around node draining.

## Motivation

> The goal of this KEP is to analyze and improve node maintenance in Kubernetes.

Reviewer 19: I don't think "node maintenance" is descriptive enough for motivation.  I would agree that most of the time when admins are doing maintenance it is for a Node.  However, a Node is just a Kubernetes API, like a Pod.  So what is to stop me for saying, "I want to do maintenance on some hardware that these 3 Pods are using"?  The same use-case applies: I want to do maintenance and Kubernetes API objects x, y, and z are blocking me from doing it (I don't want to impact users).

The motivation should be Kubernetes API agnostic.

I think a better perspective is we want to describe maintenance in Kubernetes.  For example, there are numerous Kubernetes APIs to enable replication.  Where these APIs come up short, is when you have 1 replica.  This is because there's a collision between replication requirements and maintenance requirements.  Maintenance is ideal to do when you have a low replica count (off-peak) but replication always wants to defend high-availability.  Therefore, the missing piece is we need to teach the replication APIs, and the end-user, about a new desired state: maintenance.

> Node maintenance is a request from a cluster administrator to remove all pods from a node(s) so

Reviewer 19: > Node maintenance is a request from a cluster administrator to remove all pods from a node(s)
>
I don't agree with.  We don't actually know the cluster administrators perspective.  We only know this cluster admin declared the Node API should be in "Maintenance", whatever that means.  We also don't know how the admin defines "draining" the Node.

Reviewer 20: I think this is key.

Reviewer 20: I agree with @rthallisey  here -- and I believe this is a key.

hardware or firmware upgrade, or simply to remove the node as it is no longer needed.

Kubernetes has existing support for this use case in the following way with `kubectl drain`:
1. There are running pods on node A, some of which are protected with PodDisruptionBudgets (PDB).
2. Set the node `Unschedulable` (cordon) to prevent new pods from being scheduled there.
3. Evict (default behavior) pods from node A by using the eviction API (see [kubectl drain worklflow](https://raw.githubusercontent.com/kubernetes/website/f2ef324ac22e5d9378f2824af463777182817ca6/static/images/docs/kubectl_drain.svg)).
4. Proceed with the maintenance and shut down or restart the node.
5. On platforms and nodes that support it, the kubelet will try to detect the imminent shutdown and
   then attempt to perform a [Graceful Node Shutdown](https://kubernetes.io/docs/concepts/architecture/nodes/#graceful-node-shutdown):
   - delay the shutdown pending graceful termination of remaining pods
   - terminate remaining pods in reverse priority order (see [pod-priority-graceful-node-shutdown](https://kubernetes.io/docs/concepts/architecture/nodes/#pod-priority-graceful-node-shutdown))

> The main problem is that the current approach tries to solve this in an application agnostic way

Reviewer 19: I'm going to disagree here too.  I don't think `kubectl drain` being "application agnostic" is a problem.  I think the problem is, `kubectl drain` is another Kubernetes API that doesn't understand what Maintenance is.   And the admin has no way to express what Maintenance is.

Reviewer: `kubectl drain` is not an API @rthallisey 

Draining is largely client side / isn't directly exposed as intent via the API. Hence this KEP.

Reviewer 21: Splitting hairs a bit. arguably `kubectl drain` is an api, just a pretty horrible one. Applications do code against it. :/

The goal, is to make a "proper" api for it I think.

   taint manager may prematurely terminate pods that are currently under Node Graceful Shutdown.
   This can also happen with other mechanism (e.g., different types of evictions). This has been
   discussed in the issue [kubernetes/kubernetes#124448](https://github.com/kubernetes/kubernetes/issues/124448)
   and in the issue [kubernetes/kubernetes#72129](https://github.com/kubernetes/kubernetes/issues/72129).
9. There is not enough metadata about why the node drain was requested or why the pods are
   terminating. This has been discussed in the issue [kubernetes/kubernetes#30586](https://github.com/kubernetes/kubernetes/issues/30586)
   and in the issue [kubernetes/kubernetes#116965](https://github.com/kubernetes/kubernetes/issues/116965).

Approaches and workarounds used by other projects to deal with these shortcomings:
- https://github.com/medik8s/node-maintenance-operator uses a declarative approach that tries to
  mimic `kubectl drain` (and uses kubectl implementation under the hood).
- https://github.com/kubereboot/kured performs automatic node reboots and relies on `kubectl drain`

>   implementation to achieve that.

Author: @evrardjp sure, all contributions are welcomed! We can add a more detailed description.

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

A misconfigured .spec.nodeSelector could select all the nodes (or just all master nodes) in the
cluster. This can cause the cluster to get into a degraded and unrecoverable state.

An admission plugin ([NodeMaintenance Admission](#nodemaintenance-admission)) is introduced to
issue a warning in this scenario.

## Design Details

### Kubectl

`kubectl drain`: as we can see in the [Motivation](#motivation) section, kubectl is heavily used
either manually or as a library by other projects. It is safer to keep the old behavior of this
command. However, we will deprecate it along with all the library functions. We can print a
deprecation warning when this command is used, and promote the NodeMaintenance. Additionally, pods
that support evacuation and have `evacuation.coordination.k8s.io/priority_${EVACUATOR_CLASS}`

> annotations will block eviction requests.

Reviewer 15: TBH, I think that this overlap between eviction and evacuation could be very confusing for users.
If someone defines a PDB for an application, IMO they are entitled to expect that the application is protected during drain.

> annotations will block eviction requests.

Author: > I think all tools built on top of the eviction API won't be able to drain Nodes anymore as soon as a single Pod opts into evacuation.

Not true, as the current design of the [Evacuation API](https://github.com/kubernetes/enhancements/pull/4565) stands, the eviction API will still fully work as before and will be fully supported. 

But you can opt into using NodeMaintenace+Evacuation API for more graceful pod terminations + node drain. 
We could easily enable this for example in the cluster autoscaler to observe the benefit of this feature. And it would also use eviction for pods who do not have evacuators assigned.

> annotations will block eviction requests.

Reviewer 16: My understanding is that as soon as a Pod opts into evacuation, eviction won't work anymore for this Pod ("Additionally, pods
that support evacuation and have `evacuation.coordination.k8s.io/priority_${EVACUATOR_CLASS}`
annotations will block eviction requests").

This means as soon as a single Pod in a cluster opts into evacuation, all tools that have been build on top of eviction today (Cluster API, kubectl drain, many others) won't be able to drain the Node with this Pod anymore.

// Cordon cordons all selected nodes by making them unschedulable.
Cordon NodeMaintenanceStage = "Cordon"
// Drain:
// 1. Cordons all selected nodes by making them unschedulable.
// 2. Gives an intent to drain all selected nodes by creating Evacuation objects for the
//    node's pods.
Drain NodeMaintenanceStage = "Drain"
// Complete:
// 1. Removes all Evacuation objects requested by this NodeMaintenance.
// 2. Uncordons all selected nodes by making them schedulable again, unless there is not another
//    maintenance in progress.
Complete NodeMaintenanceStage = "Complete"
)

type NodeMaintenance struct {
    ...
    Spec NodeMaintenanceSpec
    Status NodeMaintenanceStatus
}

type NodeMaintenanceSpec struct {
    // NodeSelector selects nodes for this node maintenance.
    // +required
    NodeSelector *v1.NodeSelector

    // The order of the stages is Idle -> Cordon -> Drain -> Complete.
    //
    // - The Cordon or Drain stage can be skipped by setting the stage to Complete.
    // - The NodeMaintenance object is moved to the Complete stage on deletion unless the Idle stage has been set.
    //
    // The default value is Idle.
    Stage NodeMaintenanceStage

>     // DrainPlan is executed from the first entry to the last entry during the Drain stage.

Author: > If this is a must, what if there are dependencies between Pods that don't follow this order? (e.g. a static Pod depending on a DaemonSet, or a DaemonSet depending on a Default Pod)

It is must. And I think basically all clusters follow this ordering today. Do you have an example where this is not the case?

> Similar question. Is this just a should or will the ordering based on priority per podType be enforced?

enforced

> This implies that the stages are executed based on this order and not the order of the entries. (But this would make sense if the order of the entries are enforced to have this ordering)

the order of the entries should conform to the type and priority ordering. A lower index entry with a `podSelector` and the same priority will be drained first.

>     // DrainPlan is executed from the first entry to the last entry during the Drain stage.

Reviewer 16: > It is must. And I think basically all clusters follow this ordering today. Do you have an example where this is not the case?

Sorry no concrete example. I was only wondering if it's safe to assume that the order in which Pods should be drained can simply be determined by the way that they are deployed (Deployments/operator/... vs DaemonSet). The way that Pods are deployed and drain order seem orthogonal to me.

    // DrainPlanEntry podType fields should be in the following order:
    // nil -> DaemonSet -> Static
    // DrainPlanEntry priority fields should be in ascending order for each podType.
    // If the priority and podType are the same, concrete selectors are executed first.
    //

>     // The following entries are injected into the drainPlan on the NodeMaintenance admission:

Author: Yes, by `podType`, then by `podPriority` and then by `podSelector`. As also seen in the type documentation.

    // - podPriority: 1000000000 // highest priority for user defined priority classes
    //   podType: "Default"
    // - podPriority: 2000000000 // system-cluster-critical priority class
    //   podType: "Default"
    // - podPriority: 2000001000 // system-node-critical priority class
    //   podType: "Default"
    // - podPriority: 2147483647 // maximum value
    //   podType: "Default"
    // - podPriority: 1000000000 // highest priority for user defined priority classes
    //   podType: "DaemonSet"
    // - podPriority: 2000000000 // system-cluster-critical priority class
    //   podType: "DaemonSet"
    // - podPriority: 2000001000 // system-node-critical priority class
    //   podType: "DaemonSet"
    // - podPriority: 2147483647 // maximum value
    //   podType: "DaemonSet"
    // - podPriority: 1000000000 // highest priority for user defined priority classes
    //   podType: "Static"
    // - podPriority: 2000000000 // system-cluster-critical priority class
    //   podType: "Static"

>     // - podPriority: 2000001000 // system-node-critical priority class

Author: https://github.com/kubernetes/enhancements/pull/4213#discussion_r1781854516

I would expect the apiserver to terminate last. Then the kubelet will lose the connection. Nevertheless this is a kubeadm issue, because kubelet is not connected to the load balancer. I do not see how the NodeMaintenace can help here?

>     // - podPriority: 2000001000 // system-node-critical priority class

Reviewer 16: > Nevertheless this is a kubeadm issue, because kubelet is not connected to the load balancer.

I think kubelet cannot comunicate via the loadbalancer with kube-apiservers without violating the Kubernetes skew policy (example: kubelet 1.31 is not allowed to communicate with 1.30 apiservers). The only option that kubeadm has to not violate the Kubernetes skew policy during upgrades is to communicate with the local apiserver.

    //   podType: "Static"
    // - podPriority: 2147483647 // maximum value
    //   podType: "Static"
    //
    // Duplicate entries are not allowed.
    // This field is immutable.
    DrainPlan []DrainPlanEntry

    // Reason for the maintenance.
    Reason string
}

const (
// Default selects all pods except DaemonSet and Static pods.	
Default PodType = "Default"
// DaemonSet selects DaemonSet pods.
DaemonSet PodType = "DaemonSet"
// Static selects static pods.
Static PodType = "Static"
)

> type DrainPlanEntry struct {

Reviewer 15: I have question/doubt about this API.

1) If I got it right, in order to define a rule that targets a specific set of Pods, it is required that those Pods have an identifying PodPriority. Is this correct? 

If true, what if a user needs to define a drain rule for an application where it can define a pod priority (the users not can change freely pods in applications, e.g. if the Pods are managed by some operator)?

Also, is it correct to introduce a dependency between PodPriority and drain? what if users wants to define a drain rule for an application, but they don't want that this application triggers preemption for other apps when scheduled? Are we mixing up two orthogonal problems?

2) It seems also that each rule Must be specifically defined for a single PodType, Is this correct?

if true, I'm worried that we are surfacing is API that some of the internals of the system, the fact that different K8s components takes care of different type pods.

As a user, I could find really useful to define a specific drain rule for my application, but when I try to image such rule I think at something like "I want to evict Portworx after evicting all the other pods". 

I would expect it to be a responsibility of the system to figure out what kind of Pods my application has, and do the right thing.

But if instead we expect users to dig into applications, usually designed by other teams, might be not even in my company, figure out what kind of pods the application has, and the map this to all the intricacies of K8s components and node drain, this could really impact the usability of this feature.

Reviewer: Good feedback. I'd be very happy to see an initial alpha API (either in-tree or a CRD) where we can practice triggering declarative draining, and encourage early adopters to give us their thoughts whilst the new API is still plastic (mouldable, fixable).

Reviewer: To me, declarative node maintenance is / should be much more about:
- I intend to empty this node / this rack / everything with this storage atttached to it
- here's when it will happen, and maybe here's a set of taints that are likely to get implied
  - might be good to know in case your app intends to tolerate these
- controllers are welcome to handle this gracefully before k-c-m (I assume) starts telling the kubelet to delete Pods

and out of all of that, the most key detail to capture is that it is an intent to maintain, rather than a verb masquerading as a noun (in other words, we aren't making a NodeDrainRequest or a PodRemovalSchedule).

Reviewer 15: Makes sense.
Might be the best way to bridge the two worlds is to think to a two step process/two CRDs:
- Allow application developers to declare Drain rules for their applications, like they already do for PDBs
- Allow cluster operators to declare node maintenance intent, which will then collect and use Drain rules if necessary

Also, FYI, in Cluster API we are thinking about something to make our machine drain process more flexible, and due to a few constraints, we need it in the short term; however, I can easily image how Cluster API users might benefit from something similar to what we are discussing in this thread (when it becomes available).

Author: I think the point of the DrainPlan is not to define rules on how to gracefully terminate an application. This is an admin level tool that is more coarse and basically just looks at different classes of applications (e.g. user vs system). And can differentiate between different control-plane components.

The user does not have to tinker with the priorities to achieve the desired result. They also have no guarantee that the NodeMaintenance will be the same in every cluster (e.g. all of the user app pods can fall into the same termination bucket).

To solve this, users can simply run all of their pods at e.g. priority 0 an utilize the [Evacuation API](https://github.com/kubernetes/enhancements/pull/4565) logic. Users can add an evacuator to their app pods and downscale the app according to their needs. For example, the evacuator can keep pod that provides a service alive until a pod that depends on that service terminates.

NodeMaintenance and Evacuation are intended to be extensible and for users to be able to depend on the presence of these APIs in the cluster (similar to eviction + PDB). This is the main reason why it is hard to introduce this as a CRD.

Reviewer 15: I agree declaring node maintenance is an admin operation.
But if node maintenance will only be about corse configuration (user vs system or deamon set vs static pods vs everything else), it will fail to address the friction that exists between cluster admin and application teams when a node must be drained.

WRT to the evacuation API, my personal take is that this is more a solution for advanced use cases like operators. The majority of folks need some more generic, and DrainRules have the potential to fill this gap (but in order to do so, they must be decoupled by NodeMaintenance)

Author: Yes there is a friction between the cluster admin and applicaiton teams. But it is quite a difficult task to offer a strictly declarative API to support all possible graceful draining scenarios. There could be a lot of dependencies and edge cases. I am not convinced we want to support that.

The Evacuation API allows you to do anything, but I agree that it is a quite advanced/handful for a normal user. That is why we will try to offer clever draining behaviors for some workloads (e.g. Deployment, HPA, configurable StatefulSet, etc.). And also, the Evacuation API can be used as a building block. So other components in the ecosystem can implement DrainRules and offer it as an Operator/CRD that normal users can use for a subset of applications.

Reviewer 15: Thank you for the explanation, I appreciate your patience in following my questions.

However, if what we need to solve this problem for both cluster admin and application teams is to eventually build yet another layer on top of the evacuation API, frankly speaking IMO the complexity is just too much (it is hard to figure out how m:n rules works in this proposal, let's image another layer of rules on top).

If I combine this with proposed breaking changes like draining daemonst sets and static pods, and a few other thread of discussion, it is hard to figure out how would look like the final state with a cohesive solution that helps both cluster admin and application teams, and also when this will be available.

Considering this, I'm starting to lean towards the opinion already expressed by other contributors providing feedback on this proposal that NodeMaintenance in the current form should be incubated as an external CRD, and then move to k/k when it is mature. But this depends also by how other thread are addressed.

In the meantime, I kindly ask you to add the idea discussed in this thread as alternatives in the proposal:
- decoupling drain rules definition from the node maintenance initiation (the split is what enables multiple personas to play part in this use case)
- make drain rule more application friendly (also avoiding to surface K8s internals in the user facing API)

cc @vincepri @neolit123 @sbueringer @chrischdi given that this popped up in a few SIG Cluster lifecycle related discussions

Author: > If what we need to solve this problem for both cluster admin and application teams is to eventually build yet another layer on top of the evacuation API, frankly speaking IMO the complexity is just too much (it is hard to figure out how m:n rules works in this proposal, let's image another layer of rules on top).

Another layer (whatever it is) above the Evacuation API is necessary as the Evacuation API servers as a base and implements graceful drain which is mostly not possible today.

> If I combine this with proposed breaking changes like draining daemonst sets and static pods,

Can you please let me know which part do you consider a breaking change?

> decoupling drain rules definition from the node maintenance initiation (the split is what enables multiple personas to play part in the solution of this problem)

I think it makes sense to decouple the rules. We probably still want to keep the admission logic. We could also introduce a `Default` DrainPlan that would be selected if none was supplied to the NodeMaintenance as I would expect it might be hard to select the correct one for some drainers.

> make drain rule more application friendly (also avoiding to surface K8s internals in the user facing API)

Application user facing API could be a nice followup if the community finds it useful.

    // PodSelector selects pods according to their labels.

>     // This can help to select which pods of the same priority should be evacuated first.

Reviewer 16: Would pods that match no podselector be skipped?

    // +optional

>     PodSelector *metav1.LabelSelector

Reviewer 16: Should it also be possible to select by namespace?

Reviewer: IMO: not for alpha!

When you maintain a node, you expect to take all the Pods offline or put them at risk. Let's ship something useful and then see if we want to iterate.

Reviewer 15: Begin this a global CRD, the lack of a namespace selector seems a relevant GAP for adoption

Author: What is the intended use case and benefit for the namespace? 

The main differentiator should be the pod priority. I have added the label selector to mainly differentiate between different components running for example the ` system-cluster-critical` or the `system-node-critical` priority. For example, if you are running some special etcd/apiserver configuration, you could control how the instances downscale during an upgrade. So, basically to have an additional ordering to just a priority. But in most cases the priority should be enough, and downscaling by namespaces could slow down the process considerably.

Reviewer 16: To me it seems unrealistic that ordering for Pod drain will be configured by setting Pod priorities on all Pods in a cluster, especially if not a single team has control over an entire cluster and can set Pod priorities on all relevant Pods.

We had this case here in Cluster API where folks where using Portworx which uses an operator to manage Pods: https://github.com/kubernetes-sigs/cluster-api/issues/11024 (not a DaemonSet)

I assume that if there is a use-case for a PodSelector the same probably also applies for a namespace selector (I think outside of very limited use cases it's not safe to assume Pod labels are unique across namespaces)

    // them are selected and set for that node. Thus, the DrainTargets do not have to correspond
    // to the entries in .spec.drainPlan for a single NodeMaintenance instance.
    // DrainTargets cannot backtrack and will target more pods with each update until all pods on
    // the node are targeted.
    DrainTargets  []DrainPlanEntry
    // DrainMessage may specify a state of the drain on this node and a reason why the drain
    // targets are set to a particular values.
    DrainMessage string
    // Number of pods that have not yet started evacuation.
    PodsPendingEvacuation int32
    // Number of pods that have started evacuation and have a matching Evacuation object.
    PodsEvacuating int32
}

const (
    // DrainedCondition is a condition set by the node-maintenance controller that signals
    // whether all pods pending termination have terminated on all target nodes when drain is
    // requested by the maintenance object.
    DrainedCondition = "Drained"
}
```

### NodeMaintenance Admission

`nodemaintenance` admission plugin will be introduced.

It will validate all incoming requests for CREATE, UPDATE, and DELETE operations on the

> NodeMaintenance objects. All nodes matching the `.spec.nodeSelector` must pass an authorization

Author: These are very good points. I would expect us to either
- Do additional checks on admission.
- Or just freeze the set of nodes, either on NodeMaintenance admission or when we hit the `Cordon` stage.

The 2nd options seems more feasible to me ATM.

that the cluster may get into a degraded and unrecoverable state. The warning is non-blocking and
such NodeMaintenance is still valid and can proceed.

### NodeMaintenance Controller

Node maintenance controller will be introduced and added to `kube-controller-manager`. It will
observe NodeMaintenance objects and have the following main features:

#### Idle

The controller should not touch the pods or nodes that match the selector of the NodeMaintenance
object in any way in the `Idle` stage.

#### Finalizers and Deletion of the NodeMaintenance

When a stage is not `Idle`, `nodemaintenance.k8s.io/maintenance-completion` finalizer is placed on
the NodeMaintenance object to ensure uncordon and removal of Evacuations upon deletion.

When a deletion of the NodeMaintenance object is detected, its `.spec.stage` is set to `Complete`.
The finalizer is not removed until the `Complete` stage has been completed.

#### Cordon

When a `Cordon` or `Drain` stage is detected on the NodeMaintenance object, the controller
will set (and reconcile) `.spec.Unschedulable` to `true` on all nodes that satisfy

> `.spec.nodeSelector`. It should alert via events if too many occur appear and a race to change

Reviewer 16: Is there a word missing after many?

2. Pods whose evacuators have initiated evacuation and support cancellation
   (`.status.evacuationCancellationPolicy=Allow`) should cancel the evacuation and keep the pods
   available.
3. Pods whose evacuators have initiated evacuation and do not support cancellation
   (`.status.evacuationCancellationPolicy=Forbid`) should continue the evacuation and eventually
   terminate the pods.

#### Drain

When a `Drain` stage is detected on the NodeMaintenance object, Evacuation objects are created for
selected pods ([Pod Selection](#pod-selection)).

```yaml
apiVersion: v1alpha1
kind: Evacuation
metadata:
  finalizers:
    evacuation.coordination.k8s.io/instigator_nodemaintenance.k8s.io
  name: f5823a89-e03f-4752-b013-445643b8c7a0-muffin-orders-6b59d9cb88-ks7wb
  namespace: blue-deployment
spec:
  podRef:
    name: muffin-orders-6b59d9cb88-ks7wb
    uid:  f5823a89-e03f-4752-b013-445643b8c7a0
  progressDeadlineSeconds: 1800

```

This is resolved to the following Evacuation object according to the pod on admission:

```yaml

apiVersion: v1alpha1
kind: Evacuation
metadata:
  finalizers:
    evacuation.coordination.k8s.io/instigator_nodemaintenance.k8s.io

>   labels:

Author: `app: muffin-orders`? Yes it is copied from the pod.

>   labels:

Reviewer 16: It's missing in the YAML above

    app: muffin-orders
  name: f5823a89-e03f-4752-b013-445643b8c7a0-muffin-orders-6b59d9cb88-ks7wb
  namespace: blue-deployment
spec:
  podRef:
    name: muffin-orders-6b59d9cb88-ks7wb
    uid:  f5823a89-e03f-4752-b013-445643b8c7a0
  progressDeadlineSeconds: 1800
  evacuators:

>     - evacuatorClass: deployment.apps.k8s.io

Author: There should be additional KEPs in the future. Cannot promise how many controllers this will affect and to what extent.

>     - evacuatorClass: deployment.apps.k8s.io

Reviewer 16: Sounds good. Wasn't sure if this is a future-goal or a non-goal (in the sense of we intentionally don't want to do this)

      priority: 10000
      role: controller
```

The node maintenance controller requests the removal of a pod from a node by the presence of the
Evacuation. Setting `progressDeadlineSeconds` to  1800 (30m) should give potential evacuators
enough time to recover from a disruption and continue with the graceful evacuation. If the
evacuators are unable to evacuate the pod, or if there are no evacuators, the evacuation controller
will attempt to evict these pods, until they are deleted.

The only job of the node maintenance controller is to make sure that the Evacuation object exist
and has the `evacuation.coordination.k8s.io/instigator_nodemaintenance.k8s.io` finalizer. 

#### Pod Selection

The pods for evacuation would first be selected by node (`.spec.nodeSelector`). NodeMaintenance

> should eventually remove all the pods from each node. To do this in a graceful manner, the

Reviewer 15: + 💯 to allow definition of rules allowing to intentionally skip drain for a set of Pods, it is a request that popped up several time in Cluster API, mostly due to folks deploying stuff tolerating all taints or just the ones applied by drain.

Notably, it is also worth to notice that users don't always have the chance to change Pods, e.g. because they are managed by some operator, because the app is installed by something that keep the spec in sync etc etc.
In those cases, skip drain could provide a nice exit path for operators trying to unblock node deletion.

> should eventually remove all the pods from each node. To do this in a graceful manner, the

Author: > I'm not entirely sure if the order is configurable (see questions above),
it is not

> This means that e.g. monitoring agents (logging, metrics, ...) that are usually deployed via DaemonSets will be terminated before static pods. Which in turn means that we won't have visibility during the last stages of the drain. I think this would make it hard to detect / troubleshoot the last parts of the drain.

This is similar to today: DaemonSets are also terminated before static pods.

> I'm also wondering in general if it would make sense to allow to skip drain for certain Pods:

> +💯 to add support for rules allowing to intentionally skip drain for a set of Pods, it is a request that popped up several time in Cluster API, mostly due to folks deploying stuff tolerating all taints or just the ones applied by drain.

I think the presumption is to better tune the types of workloads and their priorities. The goal of the NodeMaintenance is to remove all pods from a node and most likely to take the node offline.

> Notably, it is also worth to notice that users don't always have the chance to change Pods, e.g. because they are managed by some operator, because the app is installed by something that keep the spec in sync etc etc.
In those cases, skip drain could provide a nice exit path for operators trying to unblock node deletion.

> What about Pods that are tolerating all taints. I think this is a valid use case as sometimes folks want to enforce that Pods are running on all Nodes independent of whatever taints might be on that Node (for vendors it's impossible to tell which custom taints folks might use).
> - I think in that scenario a Node drain will be just stuck indefinitely. We are able to evict/evacuate these Pods, but the kube-scheduler will just reschedule them on the same Node.
> - Because there can be different personas involved it's also not always possible for the cluster-admin to simply modify tolerations of "third-party" Pods

These are the problems that the NodeMaintenance+Evacuation API tries to solve. Operators and any other actor should be aware of NodeMaintenance and Evacuation objects and they should offer the exit path automatically. And not block the drain and wait for manual intervention.

> should eventually remove all the pods from each node. To do this in a graceful manner, the

Reviewer 16: > This is similar to today: DaemonSets are also terminated before static pods.

Is this a part of the kubelet shutdown feature?

> controller will first ensure that lower priority pods are evacuated first for the same pod type.

Author: > If I understand correctly, this proposal assumes that the priority of Pods is also the (reverse) order in which Pods should be drained. I'm not sure if that is a correct assumption.

yes

> I wonder if it should be possible to define drain orders independent of Pod priorities.

We work with pod priorities in other places in kube and they are used by all workloads today. What would be the benefit of not using them?

> controller will first ensure that lower priority pods are evacuated first for the same pod type.

Reviewer 16: I think coupling the Pod priorities to drain order could be problematic:
1.) I think nobody today anticipates that the Pod priority is also the field that decides the drain order
  * I'm mostly not sure if it's safe the assume that the "prioritiy" of a Pod is also *always* the reverse order in which it should be drained
2.) What if folks don't have the permissions to adjust the priority of a Pod, but still want to be able to control drain order (this would easily possible with a pod & namespace selector because it doesn't require permissions to change the Pod itself)

  ...
```

The progress of the drain should not be backtracked. If an intersecting `maintenance-c` is created,
the node one progress should stay the same regardless of the node maintenance drainPlan.

```yaml
apiVersion: v1alpha1
kind: NodeMaintenance
metadata:
  name: "maintenance-c"
spec:
  nodeSelector:
    # selects nodes one and four
  stage: Drain
  drainPlan:
    - podPriority: 2000
      podType: Default
    - podPriority: 15000
      podType: Default
  ...
status:
  drainStatus:
    podsPendingEvacuation: 90
    podsEvacuating: 35
    drainMessage: "Evacuating"
    reachedDrainTargets:
      - podPriority: 2000
        podType: Default
  ...
---
apiVersion: v1
kind: Node
metadata:
  name: "one"
status:
  maintenanceStatus:
    podsPendingEvacuation: 70
    podsEvacuating: 30
    drainMessage: "Evacuating (limited by maintenance-b, maintenance-c)"
    drainTargets:
      - podPriority: 10000
        podType: Default
  ...
---
apiVersion: v1
kind: Node
metadata:
  name: "four"
status:
  maintenanceStatus:
    podsPendingEvacuation: 20
    podsEvacuating: 5
    drainMessage: "Evacuating"
    drainTargets:
      - podPriority: 2000
        podType: Default
  ...
---

> ```

Reviewer 15: +1 
It could be really help having an example that shows how a user can figure out why a node drain is blocked when multiple node maintenance apply.
(also considering all the moving parts 😅, or at least Evacuation, Eviction, DaemonSet controller, kubelet)

> ```

Author: Each Node will have its status and the NodeMaintenace should be the aggregate. Please take into account that these are not final designs and may change considerably from the final API.  There is still no consensus if there should be m:n mapping, for example.

We can add the debugging parts once we move into the PRR phase.

Unfortunately, a similar precondition is broken for the `maintenance-c`, so we can at least emit an
event saying that we are fast-forwarding `maintenance-c` due to existing older maintenance(s). In
the extreme scenario, node one may already be turned off and creating a new maintenance that
assumes priority X pods are still running will not help to bring it back. Emitting an event would
help with observability and might help cluster admins better schedule node maintenances.

##### PodTypes and Label Selectors Progression

An example progression for the following drain plan might look as follows:

```yaml
spec:
  stage: Drain
  drainPlan:
    - podPriority: 1000
      podType: Default
    - podPriority: 2000
      podType: Default
      podSelector:
        matchLabels:
          app: postgres
    - podPriority: 2147483647
      podType: Default
    - podPriority: 1000
      podType: DaemonSet
    - podPriority: 2147483647
      podType: DaemonSet
    - podPriority: 2147483647
      podType: Static
status:
  nodeStatuses:
    - nodeRef:
        name: five
      drainTargets:
        - podPriority: 1000
          podType: Default

>         - podPriority: 1000

Author: We have moved `nodeStatuses` to each node https://github.com/kubernetes/enhancements/pull/4213#discussion_r1631338976. These places require an update of that situation.

workloads where availability is paramount. Therefore, the DaemonSet controller should respond to
the Evacuation only if there is a NodeMaintenance happening on that node and the DaemonSet is in
the `drainTargets`. For example, if we observe the following NodeMaintenance:

```yaml
apiVersion: v1alpha1
kind: NodeMaintenance
...
status:
  nodeStatuses:
    - nodeRef:
        name: six
      drainTargets:
        - podPriority: 2147483647
          podType: Default
        - podPriority: 5000
          podType: DaemonSet
  ...
```

To fulfil the Evacuation API, the DaemonSet controller should register itself as a controller
evacuator. To do this, it should ensure that the following annotation is present on its own pods.

```yaml
evacuation.coordination.k8s.io/priority_daemonset.apps.k8s.io: "10000/controller"
```

The controller should respond to the Evacuation object when it observes its own class
(`daemonset.apps.k8s.io`) in `.status.activeEvacuatorClass`.

For the above node maintenance, the controller should not react to Evacuations of DaemonSet pods
with a priority greater than 5000. This state should not normally occur, as Evacuation requests

> should be coordinated with NodeMaintenance. If it does occur, we should not encourage this flow by

Author: It can be used if you implement a custom Evacuator. I think it does not make sense for the DaemonSet controller to implement it separately without the NodeMaintenance ATM.

> should be coordinated with NodeMaintenance. If it does occur, we should not encourage this flow by

Reviewer 16: It would have been nice if we would be able to evict DaemonSet Pods independent of the NodeMaintenance API if the NodeMaintenance API cannot be used.

Example, I could imagine that NodeMaintenance API will not be compatible with what we want to be able to do in Cluster API, so it would have been just nice to use it without the NodeMaintenance API.

(Context: https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20240930-machine-drain-rules.md)

Controllers can then implement the migration/termination either by reacting to the EvictionRequests
API or by reacting to the NodeMaintenance API if they need more details.

### User Stories

#### Story 1

As a cluster admin I want to have a simple interface to initiate a node drain/maintenance without
any required manual interventions. I want to have an ability to manually switch between the
maintenance phases (Planning, Cordon, Drain, Drain Complete, Maintenance Complete). I also want to
observe the node drain via the API and check on its progress. I also want to be able to discover
workloads that are blocking the node drain.

#### Story 2

As an application owner, I want to run single replica applications without disruptions and have the
ability to easily migrate the workload pods from one node to another. This also applies to
applications with larger number of replicas that prefer to surge (upscale) pods first rather than
downscale.

#### Story 3

Cluster or node autoscalers that take on the role of `kubectl drain` want to signal the intent to
drain a node using the same API and provide a similar experience to the CLI counterpart.

> 

Reviewer 21: ```suggestion

#### Story 4

Some daemonset pods can only be safely recreated after the node is drained.

There's updateStrategy OnDelete and RollingUpdate, but an OnNodeEvactuation would be really useful for some workloads, and could be enabled on top of this api.
```

    //   podType: "Default"
    // - podPriority: 2000001000 // system-node-critical priority class
    //   podType: "Default"
    // - podPriority: 2147483647 // maximum value
    //   podType: "Default"
    // - podPriority: 1000000000 // highest priority for user defined priority classes
    //   podType: "DaemonSet"
    // - podPriority: 2000000000 // system-cluster-critical priority class
    //   podType: "DaemonSet"
    // - podPriority: 2000001000 // system-node-critical priority class
    //   podType: "DaemonSet"
    // - podPriority: 2147483647 // maximum value
    //   podType: "DaemonSet"
    // - podPriority: 1000000000 // highest priority for user defined priority classes
    //   podType: "Static"
    // - podPriority: 2000000000 // system-cluster-critical priority class
    //   podType: "Static"
    // - podPriority: 2000001000 // system-node-critical priority class
    //   podType: "Static"
    // - podPriority: 2147483647 // maximum value
    //   podType: "Static"
    //
    // Duplicate entries are not allowed.
    // This field is immutable.
    DrainPlan []DrainPlanEntry

    // Reason for the maintenance.

>     Reason string

Reviewer 22: Will it be enough to use `string` here to cover all scenarios? 
For some cases it will be better to have smth "machine readable" here.

User story:
As a platform engineer I want to provide machine readable reason for the users of the platform. To empower user to write some automations based on this reason.  
For that case better to write some abstract interface here. For example - to use ObjectReference here like cluster-api do with infra-provider.
